--- XFree86-4.3.0/xc/config/cf/xfree86.cf.orig	Wed May 14 22:40:08 2003
+++ XFree86-4.3.0/xc/config/cf/xfree86.cf	Wed May 14 22:42:15 2003
@@ -1185,7 +1185,7 @@
   && !defined(ia64Architecture) \
   && !defined(s390xArchitecture) \
   && !defined (x86_64Architecture)
-#  define BuildPexExt		NO	/* But we turn off PEX anyway now */
+#  define BuildPexExt		YES	/* But we turn off PEX anyway now */
 # else
 #  define BuildPexExt		NO
 # endif
@@ -1434,7 +1434,7 @@
  */
 #ifndef BuildXIE
 # if XF86Server
-#  define BuildXIE		NO
+#  define BuildXIE		YES
 # else
 #  define BuildXIE		NO
 # endif
--- XFree86-4.2.0/xc/programs/xdpyinfo/Imakefile.orig	Thu Dec 27 21:02:34 2001
+++ XFree86-4.2.0/xc/programs/xdpyinfo/Imakefile	Sun Feb  3 10:56:30 2002
@@ -10,11 +10,6 @@
         SHMDEFINES = -DMITSHM
 #endif
 
-#if BuildXIElib
-        XIEDEFINES = -DXIE
-        XIEDEPLIBS = $(DEPXIELIB)
-           XIELIBS = $(XIELIB)
-#endif
 
 #if BuildXF86VidModeLibrary
     VIDMODEDEFINES = -DXF86VIDMODE
