diff -Nru xc.orig/programs/Xserver/hw/xfree86/xf86Version.h xc/programs/Xserver/hw/xfree86/xf86Version.h
--- xc.orig/programs/Xserver/hw/xfree86/xf86Version.h	2003-06-25 13:07:21.000000000 +0200
+++ xc/programs/Xserver/hw/xfree86/xf86Version.h	2003-06-27 14:04:29.000000000 +0200
@@ -19,7 +19,7 @@
 /* Define these for compatibility.  They'll be removed at some point. */
 #define XF86_VERSION_SUBMINOR	XF86_VERSION_PATCH
 #define XF86_VERSION_BETA	0
-#define XF87_VERSION_ALPHA	XF86_VERSION_SNAP
+#define XF86_VERSION_ALPHA	XF86_VERSION_SNAP
 
 #define XF86_VERSION_CURRENT					\
    XF86_VERSION_NUMERIC(XF86_VERSION_MAJOR,			\
