diff -urN xc.orig/programs/xkbcomp/symbols/pc/pl2 xc/programs/xkbcomp/symbols/pc/pl2
--- xc.orig/programs/xkbcomp/symbols/pc/pl2	2003-04-25 13:09:44.000000000 +0200
+++ xc/programs/xkbcomp/symbols/pc/pl2	2003-08-30 21:29:21.000000000 +0200
@@ -41,6 +41,6 @@
     key <BKSL>	{ [    oacute,     zacute,   dead_grave,   dead_breve ]	};
     key <AB03>	{ [         c,          C,         cent,    copyright ]	};
 
-    include "level3(ralt_switch_multikey)"
+    include "level3(ralt_switch)"
 };
 
