--- XFree86-4.0.2.orig/xc/programs/xkbcomp/symbols/ro	Tue Aug 15 12:51:32 2000
+++ XFree86-4.0.2/xc/programs/xkbcomp/symbols/ro	Fri Jan 19 22:03:56 2001
@@ -5,7 +5,6 @@
 //
 // Created by Cristian Gafton, <gafton@redhat.com> (C) 2000
 // 
-// $XFree86: xc/programs/xkbcomp/symbols/ro,v 3.1 2000/08/15 16:51:32 dawes Exp $
 
 partial hidden alphanumeric_keys modifier_keys
 xkb_symbols "basic" {
@@ -13,7 +12,7 @@
 
     key <TLDE> { [], [ acircumflex,     acircumflex	] };
     key <AC01> { [], [ atilde,		acircumflex	] };   
-    key <AC02> { [], [ scedilla,	Scedilla	] };
+    key <AC02> { [], [ masculine,      ordfeminine     ] };
     key <AD05> { [], [ thorn,		Thorn		] };
     key <AD08> { [], [ icircumflex,	Icircumflex	] };
 };
