--- xc/programs/xterm/Imakefile.orig	2006-06-11 03:49:18.234381592 +0200
+++ xc/programs/xterm/Imakefile	2006-06-11 03:52:43.109235880 +0200
@@ -46,6 +46,8 @@
         UTMPLIB = -lutempter
 #endif
 
+COLORS_256 = -DOPT_256_COLORS=1
+
 /*
  * Fixes to allow compile with X11R5, etc.
  */
@@ -185,7 +187,7 @@
     XKB_DEFINES = XkbClientDefines
    PATH_DEFINES = -DPROJECTROOT=$(PROJECTROOT)
         DEFINES = $(XKB_DEFINES) $(TERMCAPDEFINES) $(FEATURE_DEFINES) $(SCROLLBAR_RIGHT) $(UTF8_OPTION) $(XRFDEF) $(PATH_DEFINES) $(PUTENVDEF) $(IMAKEDEFINES) $(TRACEDEF)
-       INCLUDES = -I. -I/usr/include/ncurses $(XRFINCLUDES)
+       INCLUDES = -I. -I/usr/include/ncurses $(COLORS_256) $(XRFINCLUDES)
 
 #ifdef OS2Architecture
         MAINSRC = os2main.c
