diff -Nur XFree86-4.0.2.org/xc/config/cf/xfree86.cf XFree86-4.0.2/xc/config/cf/xfree86.cf
--- XFree86-4.0.2.org/xc/config/cf/xfree86.cf	Thu Dec 14 20:24:17 2000
+++ XFree86-4.0.2/xc/config/cf/xfree86.cf	Wed Dec 20 13:52:03 2000
@@ -435,9 +435,9 @@
 
 /* Pure PCI drivers should go first */
 # ifndef XF86CardDrivers
-#  define XF86CardDrivers	mga tdfx glint s3virge rendition tga \
-				savage nv DevelDrivers siliconmotion vga \
-				XF86OSCardDrivers XF86ExtraCardDrivers
+#define XF86CardDrivers		mga tdfx glint s3virge rendition tga \
+				nv sis neomagic i740 cirrus tseng \
+				trident chips apm fbdev ati vga v4l
 # endif
 #endif
 
@@ -520,9 +520,10 @@
 
 /* Pure PCI drivers should go first */
 # ifndef XF86CardDrivers
-#  define XF86CardDrivers	mga glint s3virge sis savage\
-				trident chips fbdev \
-				DevelDrivers vga \
+#  define XF86CardDrivers	mga glint nv tga s3virge sis savage\
+				neomagic i740 tdfx trident chips fbdev \
+				DevelDrivers vga cirrus tseng apm ati\
+				r128 ati v4l vga\
 				XF86OSCardDrivers XF86ExtraCardDrivers
 # endif
 #endif
