diff -urN XFree86-3.3.6.orig/xc/programs/Xserver/hw/xfree86/XF86Setup/Imakefile XFree86-3.3.6/xc/programs/Xserver/hw/xfree86/XF86Setup/Imakefile
--- XFree86-3.3.6.orig/xc/programs/Xserver/hw/xfree86/XF86Setup/Imakefile	Sat Apr 17 17:51:10 1999
+++ XFree86-3.3.6/xc/programs/Xserver/hw/xfree86/XF86Setup/Imakefile	Sat Mar 11 04:36:27 2000
@@ -81,7 +81,7 @@
 
  XCONFIG_DEFINES = $(EXT_DEFINES) -DXF86SETUP \
 		   '-DPROJECTROOT="$(PROJECTROOT)"' \
-		   '-DSERVER_CONFIG_FILE="$(LIBDIR)/$(XCONFIGFILE)"'
+		   '-DSERVER_CONFIG_FILE="/etc/X11/$(XCONFIGFILE)"'
 
  XF86SETUPLIBDIR = $(LIBDIR)/XF86Setup
          DEFINES = -DCARD_DATABASE_FILE='"$(CARDDBFILE)"' \
diff -urN XFree86-3.3.6.orig/xc/programs/Xserver/hw/xfree86/XF86Setup/phase5.tcl XFree86-3.3.6/xc/programs/Xserver/hw/xfree86/XF86Setup/phase5.tcl
--- XFree86-3.3.6.orig/xc/programs/Xserver/hw/xfree86/XF86Setup/phase5.tcl	Tue Oct 12 20:06:52 1999
+++ XFree86-3.3.6/xc/programs/Xserver/hw/xfree86/XF86Setup/phase5.tcl	Sat Mar 11 04:38:12 2000
@@ -38,7 +38,7 @@
 	set devid [lindex $DeviceIDs $idx]
         global Device_$devid
         set server [set Device_${devid}(Server)]
-	set linkname $Xwinhome/bin/X
+	set linkname /etc/X11/X
 	set lastlink $linkname
 	for {set nlinks 0} \
 		{[file exists $linkname] && [file type $linkname]=="link" \
@@ -75,7 +75,7 @@
 		if [string match XFCom* $server] {
 		    set serverpath $Xwinhome/bin/$server
 		} else {
-		    set serverpath $Xwinhome/bin/XF86_$server
+		    set serverpath ../../usr/X11R6/bin/XF86_$server
 		}
 		if [catch "link $serverpath X" ret] {
 		    mesg "Link creation failed!\n\
diff -urN XFree86-3.3.6.orig/xc/programs/Xserver/hw/xfree86/xf86config/xf86config.c XFree86-3.3.6/xc/programs/Xserver/hw/xfree86/xf86config/xf86config.c
--- XFree86-3.3.6.orig/xc/programs/Xserver/hw/xfree86/xf86config/xf86config.c	Mon Dec 20 17:40:07 1999
+++ XFree86-3.3.6/xc/programs/Xserver/hw/xfree86/xf86config/xf86config.c	Sat Mar 11 04:36:27 2000
@@ -107,7 +107,7 @@
 #endif
 
 /*
- * Define this to have /etc/XF86Config prompted for as the default
+ * Define this to have /etc/X11/XF86Config prompted for as the default
  * location to write the XF86Config file to.
  */
 #define PREFER_XF86CONFIG_IN_ETC
@@ -215,7 +215,7 @@
 "This program will create a basic " CONFIGNAME " file, based on menu selections you\n"
 "make.\n"
 "\n"
-"The " CONFIGNAME " file usually resides in " TREEROOTLX " or /etc. A sample\n"
+"The " CONFIGNAME " file usually resides in " TREEROOTLX " or /etc/X11. A sample\n"
 CONFIGNAME " file is supplied with XFree86; it is configured for a standard\n"
 "VGA card and monitor with 640x480 resolution. This program will ask for a\n"
 "pathname when it is ready to write the file.\n"
@@ -2986,7 +2986,7 @@
 #endif
 #ifdef PREFER_XF86CONFIG_IN_ETC
 #if !(defined(__QNX__) && !defined(__QNXNTO__))
-		printf("Shall I write it to /etc/XF86Config? ");
+		printf("Shall I write it to /etc/X11/XF86Config? ");
 #else
 		printf("Shall I write it to /etc/config/X11/XF86Config.%d? ", 
 			getnid());
@@ -2995,7 +2995,7 @@
 		printf("\n");
 		if (answerisyes(s))
 #if !(defined(__QNX__) && !defined(__QNXNTO__))
-			return "/etc/XF86Config";
+			return "/etc/X11/XF86Config";
 #else
 			/* In QNX, use XF86Config.nid for filename */
 			{
@@ -3030,11 +3030,11 @@
 #endif
   
   #ifndef PREFER_XF86CONFIG_IN_ETC
-  		printf("Shall I write it to /etc/XF86Config? ");
+  		printf("Shall I write it to /etc/X11/XF86Config? ");
 		getstring(s);
 		printf("\n");
 		if (answerisyes(s))
-			return "/etc/XF86Config";
+			return "/etc/X11/XF86Config";
 #endif
 #else /* __EMX__ */
 	{
diff -urN XFree86-3.3.6.orig/xc/programs/Xserver/os/wrapper.c XFree86-3.3.6/xc/programs/Xserver/os/wrapper.c
--- XFree86-3.3.6.orig/xc/programs/Xserver/os/wrapper.c	Fri Jul 30 15:07:09 1999
+++ XFree86-3.3.6/xc/programs/Xserver/os/wrapper.c	Sat Mar 11 04:36:27 2000
@@ -42,6 +42,9 @@
 #define XSERVER_PATH	"/usr/X11R6/bin/X"
 #endif
 
+#undef XSERVER_PATH
+#define XSERVER_PATH "/etc/X11/X"
+
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
