--- XFree86-4.1.0/xc/programs/Xserver/hw/xfree86/drivers/Makefile.orig	Tue Jul  3 07:06:21 2001
+++ XFree86-4.1.0/xc/programs/Xserver/hw/xfree86/drivers/Makefile	Mon Jul  2 21:22:18 2001
@@ -841,8 +841,8 @@
 
 MODULE_DEFINES = -DIN_MODULE -DXFree86Module
 
-SUBDIRS = mga glint nv tga s3virge sis rendition neomagic i740 tdfx savage cirrus vmware tseng trident chips apm glide fbdev i128 ati i810  ark cyrix siliconmotion vesa vga v4l
-DRIVERS = mga glint nv tga s3virge sis rendition neomagic i740 tdfx savage cirrus vmware tseng trident chips apm glide fbdev i128 ati i810  ark cyrix siliconmotion vesa vga v4l
+SUBDIRS = mga glint nv tga s3 s3virge sis rendition neomagic i740 tdfx savage cirrus vmware tseng trident chips apm glide fbdev i128 ati i810  ark cyrix siliconmotion vesa vga v4l
+DRIVERS = mga glint nv tga s3 s3virge sis rendition neomagic i740 tdfx savage cirrus vmware tseng trident chips apm glide fbdev i128 ati i810  ark cyrix siliconmotion vesa vga v4l
 
 OBJS = `cat driver.list`
 
