diff -urN xc-old/programs/Xserver/hw/xfree86/loader/elfloader.c xc/programs/Xserver/hw/xfree86/loader/elfloader.c
--- xc-old/programs/Xserver/hw/xfree86/loader/elfloader.c	2003-11-02 03:16:18.000000000 -0500
+++ xc/programs/Xserver/hw/xfree86/loader/elfloader.c	2003-11-02 03:21:53.000000000 -0500
@@ -921,7 +921,7 @@
 	    ErrorF("ELFCreateGOT() Unable to reallocate memory!!!!\n");
 	    return FALSE;
 	}
-#   if defined(linux) && defined(__ia64__) || defined(__OpenBSD__)
+#   if defined(linux) || defined(__OpenBSD__)
 	{
 	    unsigned long page_size = getpagesize();
 	    unsigned long round;
@@ -2761,10 +2761,16 @@
 	elffile->lsection[j].size = SecSize(i);
 	elffile->lsection[j].flags = flags;
 	switch (SecType(i)) {
-#ifdef __OpenBSD__
+#if defined(linux) || defined(__OpenBSD__)
 	case SHT_PROGBITS:
-	    mprotect(elffile->lsection[j].saddr, SecSize(i),
-		     PROT_READ | PROT_WRITE | PROT_EXEC);
+       {
+           unsigned long page_size = getpagesize();
+       unsigned long round;
+
+       round = (unsigned long)elffile->lsection[j].saddr & (page_size -1);
+       mprotect( (char *)elffile->lsection[j].saddr - round,
+            SecSize(i) + round, PROT_READ|PROT_WRITE|PROT_EXEC);
+       }
 	    break;
 #endif
 	case SHT_SYMTAB:
@@ -2959,7 +2965,7 @@
 	ErrorF("Unable to allocate ELF sections\n");
 	return NULL;
     }
-#  if defined(linux) && defined(__ia64__) || defined(__OpenBSD__)
+#  if defined(linux) || defined(__OpenBSD__)
     {
 	unsigned long page_size = getpagesize();
 	unsigned long round;
