--- XFree86-4.0.1/xc/programs/xkbcomp/rules/sgi.lst.learn-to-spell	Mon Aug 14 15:35:41 2000
+++ XFree86-4.0.1/xc/programs/xkbcomp/rules/sgi.lst	Mon Aug 14 15:36:45 2000
@@ -38,7 +38,7 @@
   jp			Japanese
   no			Norwegian
   pl			Polish
-  pt			Portugese
+  pt			Portuguese
   ru			Russian
   se			Swedish
   th			Thai
--- XFree86-4.0.1/xc/programs/xkbcomp/rules/sun.lst.learn-to-spell	Mon Aug 14 15:35:47 2000
+++ XFree86-4.0.1/xc/programs/xkbcomp/rules/sun.lst	Mon Aug 14 15:36:54 2000
@@ -27,7 +27,7 @@
   jp			Japanese
   no			Norwegian
   pl			Polish
-  pt			Portugese
+  pt			Portuguese
   ru			Russian
   se			Swedish
   th			Thai
