--- xc/programs/Xserver/hw/xfree86/fbdevhw/fbdevhw.c.orig	2006-03-16 17:50:24.000000000 +0100
+++ xc/programs/Xserver/hw/xfree86/fbdevhw/fbdevhw.c	2006-07-19 21:10:52.676789250 +0200
@@ -1,5 +1,7 @@
 /* $XFree86: xc/programs/Xserver/hw/xfree86/fbdevhw/fbdevhw.c,v 1.35 2006/03/16 16:50:24 dawes Exp $ */
 
+#include <asm/page.h>	/* #define for PAGE_* */
+
 /* all driver need this */
 #include "xf86.h"
 #include "xf86_OSproc.h"
@@ -14,8 +16,6 @@
 #include "fbdevhw.h"
 #include "fbpriv.h"
 
-#include "asm/page.h"	/* #define for PAGE_* */
-
 #include "globals.h"
 #define DPMS_SERVER
 #include <X11/extensions/dpms.h>
