diff -urN aaa/xc/programs/Xserver/hw/xfree86/ddc/xf86DDC.c xc/programs/Xserver/hw/xfree86/ddc/xf86DDC.c
--- aaa/xc/programs/Xserver/hw/xfree86/ddc/xf86DDC.c	2003-07-25 13:04:20.000000000 +0200
+++ xc/programs/Xserver/hw/xfree86/ddc/xf86DDC.c	2003-07-29 22:09:08.000000000 +0200
@@ -333,7 +333,7 @@
     
     xf86LoaderReqSymLists(i2cSymbols, NULL);
 
-    if (!dev = xf86I2CFindDev(pBus, 0x00A0)) {
+    if (!(dev = xf86I2CFindDev(pBus, 0x00A0))) {
 	dev = xf86CreateI2CDevRec();
 	dev->DevName = "ddc2";
 	dev->SlaveAddr = 0xA0;
