--- XFree86-4.2.1/xc/programs/Xserver/hw/xfree86/drivers/sis/sis_accel.c.maxxfbmem-fixup	Sun Aug 25 19:26:12 2002
+++ XFree86-4.2.1/xc/programs/Xserver/hw/xfree86/drivers/sis/sis_accel.c	Sun Aug 25 19:27:15 2002
@@ -133,7 +133,7 @@
     else 
         offset = 0;
     
-    topFB = (pSiS->maxxfbmem >= (pSiS->FbMapSize - offset)) ?
+    topFB = (pSiS->maxxfbmem < (pSiS->FbMapSize - offset)) ?
 	pSiS->maxxfbmem : pSiS->FbMapSize - offset;
     AvailFBArea.y2 = (topFB) / (pScrn->displayWidth *
                       pScrn->bitsPerPixel / 8);
--- XFree86-4.2.1/xc/programs/Xserver/hw/xfree86/drivers/sis/sis530_accel.c.maxxfbmem-fixup	Sun Aug 25 19:26:39 2002
+++ XFree86-4.2.0/xc/programs/Xserver/hw/xfree86/drivers/sis/sis530_accel.c	Sun Aug 25 19:26:56 2002
@@ -135,7 +135,7 @@
 	else if (pSiS->TurboQueue || pSiS->HWCursor) 
 	    offset = 32768;
     }
-    topFB = (pSiS->maxxfbmem >= pSiS->FbMapSize - offset) ?
+    topFB = (pSiS->maxxfbmem < pSiS->FbMapSize - offset) ?
 	pSiS->maxxfbmem : pSiS->FbMapSize - offset;
 
     /* CPU To screen color expansion indirect method */
