diff -urN XFree86-3.3.6.orig/xc/config/cf/xf86site.def XFree86-3.3.6/xc/config/cf/xf86site.def
--- XFree86-3.3.6.orig/xc/config/cf/xf86site.def	Sat Mar 11 04:34:04 2000
+++ XFree86-3.3.6/xc/config/cf/xf86site.def	Sat Mar 11 05:02:00 2000
@@ -93,8 +93,8 @@
  * Some Linux releases don't have a libtermcap.  In this case you may need
  * to uncomment the following
  *
-#define TermcapLibrary		-lncurses
  */
+#define TermcapLibrary		-lncurses
 
 /*
  * If you have Tk (which is required to build XF86Setup), uncomment this
diff -urN XFree86-3.3.6.orig/xc/programs/xterm/Imakefile XFree86-3.3.6/xc/programs/xterm/Imakefile
--- XFree86-3.3.6.orig/xc/programs/xterm/Imakefile	Sat Mar 11 05:00:52 2000
+++ XFree86-3.3.6/xc/programs/xterm/Imakefile	Sat Mar 11 05:02:00 2000
@@ -83,7 +83,7 @@
 		  -DOSMINORVERSION=$(OSMINORVERSION)
    MISC_DEFINES = /* -DALLOWLOGGING -DALLOWLOGFILEEXEC */ 
     XKB_DEFINES = XkbClientDefines
-        DEFINES = -I. $(XKB_DEFINES) $(TERMCAPDEFINES) $(FEATURE_DEFINES) $(SCROLLBAR_RIGHT)
+        DEFINES = -I. -I/usr/include/ncurses $(XKB_DEFINES) $(TERMCAPDEFINES) $(FEATURE_DEFINES) $(SCROLLBAR_RIGHT)
 
 #ifdef OS2Architecture
           MAINSRC = os2main.c
