--- xc/programs/xterm/Imakefile.orig	2013-02-15 18:40:32.585207554 +0100
+++ xc/programs/xterm/Imakefile	2013-02-15 18:41:06.161872812 +0100
@@ -50,6 +50,8 @@
         UTMPLIB = -lutempter
 #endif
 
+COLORS_256 = -DOPT_256_COLORS=1
+
 /*
  * Fixes to allow compile with X11R5, etc.
  */
@@ -214,7 +216,7 @@
         DEFINES = $(XKB_DEFINES) $(TERMCAPDEFINES) $(FEATURE_DEFINES) \
                   $(SCROLLBAR_RIGHT) $(UTF8_OPTION) $(XRFDEF) $(PATH_DEFINES) \
                   $(XTERM_FEATURE_OPTIONS) $(PUTENVDEF) \
-                  $(IMAKEDEFINES) $(TRACEDEF)
+                  $(IMAKEDEFINES) $(TRACEDEF) $(COLORS_256)
        INCLUDES = $(OBSOLETE_INCLUDES) $(XRFINCLUDES) -I/usr/include/ncurses
 
 #ifdef OS2Architecture
