--- xc/programs/Xserver/os/Imakefile.orig	2006-06-11 17:00:33.615026696 +0200
+++ xc/programs/Xserver/os/Imakefile	2006-06-11 17:02:12.674967296 +0200
@@ -193,6 +193,7 @@
 		  -I$(SERVERSRC)/Xext -I$(SERVERSRC)/render \
 		  -I$(SERVERSRC)/lbx -I$(SERVERSRC)/Xprint \
 		  -I$(TOP)/lib/Xau Krb5Includes
+ EXTRA_DEFINES = -DUSE_PAM
  DEPEND_DEFINES = $(DBM_DEFINES) $(XDMCP_DEFINES) \
 		  $(TRANS_INCLUDES) $(CONNECTION_FLAGS) DependDefines
        LINTLIBS = ../dix/llib-ldix.ln
