--- XFree86-4.3.99.15/xc/lib/Xfontcache/Imakefile.orig	2003-11-02 05:43:16.000000000 +0100
+++ XFree86-4.3.99.15/xc/lib/Xfontcache/Imakefile	2003-11-03 07:58:44.399216056 +0100
@@ -36,7 +36,7 @@
 MANSUFFIX=$(LIBMANSUFFIX)
 InstallManPage(LibName,$(LIBMANDIR))
 #if ExpandManNames
-InstallManPageAliases(LibName,$(MANDIR),FontCacheQueryExtension FontCacheQueryVersion FontCacheGetCacheSettings FontCacheChangeCacheSettings FontCacheGetCacheStatistics)
+InstallManPageAliases(LibName,$(LIBMANDIR),FontCacheQueryExtension FontCacheQueryVersion FontCacheGetCacheSettings FontCacheChangeCacheSettings FontCacheGetCacheStatistics)
 #endif
 
 DependTarget()
