--- XFree86-4.2.0/xc.old/extras/X-TrueType/xttfuncs.c	Tue Dec 18 12:23:55 2001
+++ XFree86-4.2.0/xc/extras/X-TrueType/xttfuncs.c	Tue Apr  2 23:31:14 2002
@@ -1100,7 +1100,7 @@
                                 char_width);
                 }
 
-            if (!tmpchar->characterWidth)
+            if (!tmpchar || !tmpchar->characterWidth)
                 continue;
 
                 adjust_min_max(&minchar, &maxchar, tmpchar);
