diff -urN XFree86-4.3.99.6.org/xc/programs/Xserver/hw/xfree86/os-support/linux/lnx_io.c XFree86-4.3.99.6/xc/programs/Xserver/hw/xfree86/os-support/linux/lnx_io.c
--- XFree86-4.3.99.6.org/xc/programs/Xserver/hw/xfree86/os-support/linux/lnx_io.c	2003-06-18 19:41:00.000000000 +0200
+++ XFree86-4.3.99.6/xc/programs/Xserver/hw/xfree86/os-support/linux/lnx_io.c	2003-06-18 19:41:18.000000000 +0200
@@ -66,6 +66,7 @@
 /* kbd rate stuff based on kbdrate.c from Rik Faith <faith@cs.unc.edu> et.al.
  * from util-linux-2.9t package */
 
+#include <linux/version.h>
 #include <linux/kd.h>
 #ifdef __sparc__
 #include <asm/param.h>
