--- XFree86-4.3.99.902/LinuxDriver/2D/savage_video.c.orig	2003-12-24 15:32:46.000000000 +0100
+++ XFree86-4.3.99.902/LinuxDriver/2D/savage_video.c	2003-12-24 20:04:15.781060984 +0100
@@ -40,7 +40,10 @@
 typedef unsigned long long __u64;
 #endif
 
+/* don't get system <linux/videodev2.h>, it's not sufficient */
+#define __LINUX_VIDEODEV2_H
 #include <linux/videodev.h>
+#undef  __LINUX_VIDEODEV2_H
 #include <linux/fcntl.h> /* for open flags*/
 
 /* v4l2 not in kernel's videodev.h*/
