--- XFree86-3.3.3.1/xc/programs/Xserver/hw/xfree86/XF86Setup/filelist.tcl~	Mon Nov 16 14:13:41 1998
+++ XFree86-3.3.3.1/xc/programs/Xserver/hw/xfree86/XF86Setup/filelist.tcl	Sat Jun 19 16:19:34 1999
@@ -48,8 +48,8 @@
 	lib/X11/fonts/misc/fonts.dir		444
 	lib/X11/fonts/misc/fonts.alias		444
 	lib/X11/fonts/misc/6x13.pc*		444
-	lib/X11/fonts/75dpi/fonts.dir		444
-	lib/X11/fonts/75dpi/symb10.pc*		444
+	lib/X11/fonts/*dpi/fonts.dir		444
+	lib/X11/fonts/*dpi/symb10.pc*		444
 }
 
 array set FilePermsDB {
