This patch by Branden Robinson.

--- XFree86-4.1.0/xc/lib/Xft/Imakefile~	Fri Apr 27 14:55:22 2001
+++ XFree86-4.1.0/xc/lib/Xft/Imakefile	Mon Jun  4 01:20:55 2001
@@ -22,8 +22,7 @@
 #define XftType1Dir $(LIBDIR)/fonts/Type1
 #endif
 
-XFTLIBDIR=XftLibDir
-XFTCONFIG=$(XFTLIBDIR)/XftConfig
+XFTCONFIG=$(CONFDIR)/XftConfig
 
 TYPE1DIR=XftType1Dir
 CONFIG_DEFS=-DXFT_TYPE1_DIR=\"$(TYPE1DIR)\"
@@ -68,9 +67,9 @@
 CppFileTarget(XftConfig,XftConfig.cpp,$(CONFIG_DEFS),$(ICONFIGFILES))
 
 #if InstallFSConfig
-InstallNonExecFile(XftConfig,$(XFTLIBDIR))
+InstallNonExecFile(XftConfig,$(CONFDIR))
 #else
-InstallNonExecFileNoClobber(XftConfig,$(XFTLIBDIR))
+InstallNonExecFileNoClobber(XftConfig,$(CONFDIR))
 #endif
 
 LinkConfFileLong(XftConfig,XftConfig,$(XFTLIBDIR),$(CONFDIR))
