--- XFree86-4.1.0.org/xc/Makefile	Wed Mar 22 22:23:30 2000
+++ XFree86-4.1.0/xc/Makefile	Wed Dec 20 13:41:02 2000
@@ -8,10 +8,10 @@
 # or add "MAKE = make" to this file.
 
 RELEASE = "Release 6.5"
-SHELL = /bin/sh
+SHELL = /bin/sh -e
 RM = rm -f
 MV = mv
-WORLDOPTS = -k
+WORLDOPTS =
 WIN32WORLDOPTS = -i
 TOP = .
 CURRENT_DIR = .
diff -Nur XFree86-4.0.2.org/xc/config/cf/Imake.cf XFree86-4.0.2/xc/config/cf/Imake.cf
--- XFree86-4.0.2.org/xc/config/cf/Imake.cf	Mon Nov  6 23:53:47 2000
+++ XFree86-4.0.2/xc/config/cf/Imake.cf	Wed Dec 20 13:41:02 2000
@@ -689,6 +689,10 @@
 #   define s390Architecture
 #  undef __s390__
 # endif /* s390 */
+# ifdef alpha
+#  define AlphaArchitecture
+#  undef alpha
+# endif /* alpha */
 # ifdef __alpha
 #  define AlphaArchitecture
 #  undef __alpha
diff -Nur XFree86-4.0.2.org/xc/config/cf/Imake.tmpl XFree86-4.0.2/xc/config/cf/Imake.tmpl
--- XFree86-4.0.2.org/xc/config/cf/Imake.tmpl	Tue Dec  5 01:45:28 2000
+++ XFree86-4.0.2/xc/config/cf/Imake.tmpl	Wed Dec 20 13:41:02 2000
@@ -220,7 +220,7 @@
 
 #ifndef BuildHtmlManPages
 #ifdef RmanCmd
-#define BuildHtmlManPages YES
+#define BuildHtmlManPages NO
 #else
 #define BuildHtmlManPages NO
 #endif
@@ -686,11 +686,7 @@
 #define LintlibDir $(USRLIBDIR)/lint
 #endif
 #ifndef SystemManDirectory
-#if SystemV4
 #define SystemManDirectory /usr/share/man
-#else
-#define SystemManDirectory /usr/man
-#endif
 #endif
 #ifndef ManDirectoryRoot
 #ifdef ProjectRoot
@@ -817,7 +813,7 @@
 #define NoRConst NO /* YES if const for structs of funcs is bad */
 #endif
 #ifndef InstPgmFlags
-#define InstPgmFlags -s
+#define InstPgmFlags
 #endif
 #ifndef InstBinFlags
 #define InstBinFlags -m 0755
@@ -958,7 +954,7 @@
 #define CppCmd /lib/cpp
 #endif
 #ifndef RawCppCmd
-#define RawCppCmd CppCmd -undef
+#define RawCppCmd CppCmd
 #endif
 #ifndef CppNoLineInfoOption
 #define CppNoLineInfoOption /**/
diff -Nur XFree86-4.0.2.org/xc/config/cf/X11.tmpl XFree86-4.0.2/xc/config/cf/X11.tmpl
--- XFree86-4.0.2.org/xc/config/cf/X11.tmpl	Sat Dec 16 07:39:25 2000
+++ XFree86-4.0.2/xc/config/cf/X11.tmpl	Wed Dec 20 13:41:02 2000
@@ -882,8 +882,12 @@
 #endif
 #endif
 
+#ifndef EtcX11Directory
+#define EtcX11Directory /etc/X11
+#endif
+
 #ifndef DefaultFSConfigFile
-#define DefaultFSConfigFile $(LIBDIR)/fs/config
+#define DefaultFSConfigFile $(ETCX11DIR)/fs/config
 #endif
 
 /* X-TrueType settings */
@@ -1002,7 +1006,7 @@
 #define ConfigDir $(LIBDIR)/config
 #endif
 #ifndef DocDir
-#define DocDir $(LIBDIR)/doc
+#define DocDir $(LIBDIR)/share/doc
 #endif
 #ifndef DocHtmlDir
 #define DocHtmlDir $(DOCDIR)/html
@@ -1011,31 +1015,31 @@
 #define DocPsDir $(DOCDIR)/PostScript
 #endif
 #ifndef FontDir
-#define FontDir $(LIBDIR)/fonts
+#define FontDir /usr/share/fonts
 #endif
 #ifndef AdmDir
-#define AdmDir /usr/adm
+#define AdmDir /var/log
 #endif
 #ifndef XinitDir
-#define XinitDir $(LIBDIR)/xinit
+#define XinitDir $(ETCX11DIR)/xinit
 #endif
 #ifndef XdmDir
-#define XdmDir $(LIBDIR)/xdm
+#define XdmDir $(ETCX11DIR)/xdm
 #endif
 #ifndef XdmVarDir
 #define XdmVarDir $(VARLIBDIR)/xdm
 #endif
 #ifndef XdmAuthDir
-#define XdmAuthDir $(LIBDIR)/xdm
+#define XdmAuthDir $(ETCX11DIR)/xdm
 #endif
 #ifndef XdmConfigurationSubdirectory		/* under xdm sources */
 #define XdmConfigurationSubdirectory config
 #endif
 #ifndef TwmDir
-#define TwmDir $(LIBDIR)/twm
+#define TwmDir $(ETCX11DIR)/twm
 #endif
 #ifndef XsmDir
-#define XsmDir $(LIBDIR)/xsm
+#define XsmDir $(ETCX11DIR)/xsm
 #endif
 #ifndef NlsDir
 #define NlsDir $(LIBDIR)/nls
@@ -1047,23 +1051,20 @@
 #define PexApiDir $(LIBDIR)/PEX
 #endif
 #ifndef LbxproxyDir
-#define LbxproxyDir $(LIBDIR)/lbxproxy
+#define LbxproxyDir $(ETCX11DIR)/lbxproxy
 #endif
 #ifndef LbxproxyConfigSubdir		/* under lbxproxy sources */
 #define LbxproxyConfigSubdir config
 #endif
 #ifndef ProxyManagerDir
-#define ProxyManagerDir $(LIBDIR)/proxymngr
+#define ProxyManagerDir $(ETCX11DIR)/proxymngr
 #endif
 #ifndef ServerConfigDir
-#define ServerConfigDir $(LIBDIR)/xserver
+#define ServerConfigDir $(ETCX11DIR)/xserver
 #endif
 #ifndef XPrintDir
 #define XPrintDir ServerConfigDir
 #endif
-#ifndef EtcX11Directory
-#define EtcX11Directory /etc/X11
-#endif
 #ifndef InstallSecurityConfig
 #define InstallSecurityConfig YES	/* Install sample security policy */
 #endif
@@ -1095,10 +1096,10 @@
 #define OsNameDefines /* as nothing */
 #endif
 #ifndef DefaultUserPath			/* for xdm */
-#define DefaultUserPath :/bin:/usr/bin:$(BINDIR):/usr/ucb
+#define DefaultUserPath /usr/local/bin:/bin:/usr/bin:$(BINDIR)
 #endif
 #ifndef DefaultSystemPath		/* for xdm */
-#define DefaultSystemPath /etc:/bin:/usr/bin:$(BINDIR):/usr/ucb
+#define DefaultSystemPath /sbin:/bin:/usr/sbin:/usr/bin:$(BINDIR)
 #endif
 #ifndef DefaultSystemShell		/* for xdm to set in SHELL */
 #define DefaultSystemShell BourneShell
@@ -1211,7 +1212,7 @@
 
      HTMLINDEXCMD = HtmlIndexCmd
 
-       DOCUTILSRC = $(XTOP)/doc/util
+       DOCUTILSRC = $(XTOP)/share/doc/util
         CLIENTSRC = $(TOP)/clients
           DEMOSRC = $(TOP)/demos
        XDOCMACROS = $(DOCUTILSRC)/macros.t
@@ -1225,7 +1226,7 @@
         SERVERSRC = $(XTOP)/programs/Xserver
        CONTRIBSRC = ContribDir
    UNSUPPORTEDSRC = UnsupportedDir
-           DOCSRC = $(XTOP)/doc
+           DOCSRC = $(XTOP)/share/doc
            RGBSRC = $(XTOP)/programs/rgb
       BDFTOPCFSRC = $(PROGRAMSRC)/bdftopcf
      MKFONTDIRSRC = $(PROGRAMSRC)/mkfontdir
diff -Nur XFree86-4.0.2.org/xc/config/cf/linux.cf XFree86-4.0.2/xc/config/cf/linux.cf
--- XFree86-4.0.2.org/xc/config/cf/linux.cf	Thu Dec 14 20:24:08 2000
+++ XFree86-4.0.2/xc/config/cf/linux.cf	Wed Dec 20 13:45:03 2000
@@ -286,7 +286,7 @@
  * performance.
  */
 #if BuildXF86DRI
-#  define BuildLibGlxWithoutPIC	YES
+#  define BuildLibGlxWithoutPIC	NO
 #endif
 
 /* The DRM module requires kernel services that appeared in late 2.1.x
@@ -298,7 +298,7 @@
 #  if defined(i386Architecture) || defined(ia64Architecture)
 #    if OSMajorVersion == 2
        /* Only tested for Linux 2.2.0 through 2.2.12 */
-#      if OSMinorVersion == 2 && OSTeenyVersion <= 12
+#      if OSMinorVersion == 2
 #        define BuildXF86DRM	YES
 #      else
          /* Only tested for Linux 2.3.0 through 2.3.18 */
@@ -565,7 +565,7 @@
 #define StandardCppDefines	StandardDefines
 
 #define HasVarRun		YES
-#define VarDbDirectory		$(VARDIR)/lib
+#define VarDbDirectory		/var/lib
 
 #ifndef XF86OSCardDrivers
 # define XF86OSCardDrivers	v4l
@@ -704,10 +704,10 @@
 #endif
 
 #ifndef LinkGLToUsrLib
-# define LinkGLToUsrLib		YES
+# define LinkGLToUsrLib		NO
 #endif
 # ifndef LinkGLToUsrInclude
-# define LinkGLToUsrInclude	YES
+# define LinkGLToUsrInclude	NO
 #endif
 
 
diff -Nur XFree86-4.0.2.org/xc/config/cf/site.def XFree86-4.0.2/xc/config/cf/site.def
--- XFree86-4.0.2.org/xc/config/cf/site.def	Tue Jun 27 02:55:45 2000
+++ XFree86-4.0.2/xc/config/cf/site.def	Wed Dec 20 13:41:02 2000
@@ -131,10 +131,14 @@
 #define HasSentinel	YES
 */
 
-/*
 #undef DefaultUserPath
-#define DefaultUserPath /bin:/usr/bin:$(BINDIR):/usr/ucb:/usr/local/bin
-*/
+#define DefaultUserPath /usr/local/bin:/bin:/usr/bin:$(BINDIR)
+#undef DefaultSystemPath
+#define DefaultSystemPath /sbin:/bin:/usr/sbin:/usr/bin:$(BINDIR)
+
+
+#define HasPam    YES
+#define HasPamMisc    YES
 
 
 /* You do NOT need SetUID if you only run the server under xdm */
diff -Nur XFree86-4.0.2.org/xc/config/cf/xf86site.def XFree86-4.0.2/xc/config/cf/xf86site.def
--- XFree86-4.0.2.org/xc/config/cf/xf86site.def	Thu Dec 14 20:24:16 2000
+++ XFree86-4.0.2/xc/config/cf/xf86site.def	Wed Dec 20 13:41:03 2000
@@ -134,16 +134,16 @@
  * Some Linux releases don't have a libtermcap.  In this case you may need
  * to uncomment the following
  *
-#define TermcapLibrary		-lncurses
  */
+#define TermcapLibrary		-lncurses
 
 /*
  * If you have Tk (which is required to build XF86Setup), uncomment this
  * Note: version 4.0 or 4.1 is required, and XF86Setup links it statically by
  * default.
  *
-#define HasTk			YES
  */
+#define HasTk			YES
 
 /*
  * Set the paths and names for your Tk library if they don't match the
@@ -156,25 +156,25 @@
  *
  * Some Linux releases need TkLibDir set to /usr/X11R6/lib
  *
-#define TkLibDir		/usr/local/lib
-#define TkIncDir		/usr/local/include
-#define TkLibName		tk
  */
+#define TkLibDir		/usr/lib
+#define TkIncDir		/usr/include
+#define TkLibName		tk
 
 /*
  * By default, XF86Setup is linked with a static Tk library.  To change
  * that, uncomment the following line.
  *
-#define XF86SetupUsesStaticTk	NO
  */
+#define XF86SetupUsesStaticTk	NO
 
 /*
  * If you have Tcl (which is required to build XF86Setup), uncomment this
  * Note: version 7.4 or 7.5 is required, and XF86Setup links it statically by
  * default.
  *
-#define HasTcl			YES
  */
+#define HasTcl			YES
 
 /*
  * Set the paths and names for your Tcl library if they don't match the
@@ -184,17 +184,17 @@
  *
  * Some Linux releases need TclIncDir set to /usr/include/tcl.
  *
-#define TclLibDir		/usr/local/lib
-#define TclIncDir		/usr/local/include
-#define TclLibName		tcl
  */
+#define TclLibDir		/usr/lib
+#define TclIncDir		/usr/include
+#define TclLibName		tcl
 
 /*
  * By default, XF86Setup is linked with a static Tcl library.  To change
  * that, uncomment the following line.
  *
-#define XF86SetupUsesStaticTcl	NO
  */
+#define XF86SetupUsesStaticTcl	NO
 
 /*
  * Build a server that dynamically loads the modules by setting
@@ -436,8 +436,8 @@
 /*
  * To not build/install the Cyrillic fonts, uncomment the following
  *
-#define BuildCyrillicFonts	NO
  */
+#define BuildCyrillicFonts	YES
 
 /*
  * To not build/install the Latin2 fonts, uncomment the following
@@ -626,8 +626,8 @@
  * To forceably build static libraries in addition to shared libraries,
  * uncomment this.
  *
-#define ForceNormalLib		YES
  */
+#define ForceNormalLib		YES
 
 /*
  * Uncomment this if your default tools (eg, gcc, ld, as, etc) are
@@ -713,26 +713,26 @@
  * HasGlide2 to YES and set the Glide2IncDir variable. 
  * HasGlide2 is per default NO.
  *
-#define HasGlide2 YES
  */
+#define HasGlide2 YES
 
 /*
  * Set the path to your Glide include files.
  *
-#define Glide2IncDir /usr/include/glide
  */
+#define Glide2IncDir /usr/include/glide
 
 /* 
  * The 3Dfx 3D DRI driver requires glide 3.
  *
-#define HasGlide3 YES
  */
+#define HasGlide3 YES
 
 /*
  * Set the path to your Glide 3 include files.
  *
-#define Glide3IncDir /usr/include/glide3
  */
+#define Glide3IncDir /usr/include/glide3
 
 /*
  * Undefine the following if you don't want to have config files and
diff -Nur XFree86-4.0.2.org/xc/config/cf/xfree86.cf XFree86-4.0.2/xc/config/cf/xfree86.cf
--- XFree86-4.0.2.org/xc/config/cf/xfree86.cf	Thu Dec 14 20:24:17 2000
+++ XFree86-4.0.2/xc/config/cf/xfree86.cf	Wed Dec 20 13:52:03 2000
@@ -28,7 +28,7 @@
 #define XConfigFile		XF86Config
 #endif
 #ifndef XConfigDir
-#define XConfigDir		$(LIBDIR)
+#define XConfigDir		$(ETCX11DIR)
 #endif
 #ifndef XLogFile
 #define XLogFile		XFree86
@@ -1035,7 +1035,7 @@
  * Build SuperProbe
  */
 #ifndef BuildSuperProbe
-# define BuildSuperProbe 		NO
+# define BuildSuperProbe 		YES
 #endif
 
 /*
diff -Nur XFree86-4.0.2.org/xc/config/imake/imake.c XFree86-4.0.2/xc/config/imake/imake.c
--- XFree86-4.0.2.org/xc/config/imake/imake.c	Wed Nov 29 17:33:50 2000
+++ XFree86-4.0.2/xc/config/imake/imake.c	Wed Dec 20 13:41:03 2000
@@ -952,6 +952,7 @@
   static char* yast = "/sbin/YaST";
   static char* redhat = "/etc/redhat-release";
   static char* debian = "/etc/debian_version";
+  static char* pld = "/etc/pld-release";
 
   fprintf (inFile, "%s\n", "#define LinuxUnknown    0");
   fprintf (inFile, "%s\n", "#define LinuxSuSE       1");
@@ -966,6 +967,7 @@
   fprintf (inFile, "%s\n", "#define LinuxTurbo      10");
   fprintf (inFile, "%s\n", "#define LinuxWare       11");
   fprintf (inFile, "%s\n", "#define LinuxYggdrasil  12");
+  fprintf (inFile, "%s\n", "#define LinuxPLD        13");
 
   if (lstat (yast, &sb) == 0) {
     fprintf (inFile, "%s\n", "#define DefaultLinuxDistribution LinuxSuSE");
@@ -982,6 +984,11 @@
     fprintf (inFile, "%s\n", "#define DefaultLinuxDistName Debian");
     /* You could also try to get the version of the Debian distrib by looking
      * at the content of /etc/debian_version */
+    return;
+  }
+  if (lstat (pld, &sb) == 0) {
+    fprintf (inFile, "%s\n", "#define DefaultLinuxDistribution LinuxPLD");
+    fprintf (inFile, "%s\n", "#define DefaultLinuxDistName PLD");
     return;
   }
   /* what's the definitive way to tell what any particular distribution is? */
diff -Nur XFree86-4.0.2.org/xc/programs/Xserver/hw/xfree86/XF86Setup/Imakefile XFree86-4.0.2/xc/programs/Xserver/hw/xfree86/XF86Setup/Imakefile
--- XFree86-4.0.2.org/xc/programs/Xserver/hw/xfree86/XF86Setup/Imakefile	Sat Sep  4 15:04:28 1999
+++ XFree86-4.0.2/xc/programs/Xserver/hw/xfree86/XF86Setup/Imakefile	Wed Dec 20 13:41:03 2000
@@ -118,7 +118,7 @@
 
  XCONFIG_DEFINES = $(EXT_DEFINES) -DXF86SETUP \
 		   '-DPROJECTROOT="$(PROJECTROOT)"' \
-		   '-DSERVER_CONFIG_FILE="$(LIBDIR)/$(XCONFIGFILE)"' \
+		   '-DSERVER_CONFIG_FILE="/etc/X11/$(XCONFIGFILE)"' \
 		   $(SITE_FONT_PATH) $(SITE_RGB_DB) \
 		   -DNO_WRAPPERS
 
diff -Nur XFree86-4.0.2.org/xc/programs/Xserver/hw/xfree86/XF86Setup/phase5.tcl XFree86-4.0.2/xc/programs/Xserver/hw/xfree86/XF86Setup/phase5.tcl
--- XFree86-4.0.2.org/xc/programs/Xserver/hw/xfree86/XF86Setup/phase5.tcl	Sun Apr  5 18:15:52 1998
+++ XFree86-4.0.2/xc/programs/Xserver/hw/xfree86/XF86Setup/phase5.tcl	Wed Dec 20 13:41:03 2000
@@ -38,7 +38,7 @@
 	set devid [lindex $DeviceIDs $idx]
         global Device_$devid
         set server [set Device_${devid}(Server)]
-	set linkname $Xwinhome/bin/X
+	set linkname /etc/X11/X
 	set lastlink $linkname
 	for {set nlinks 0} \
 		{[file exists $linkname] && [file type $linkname]=="link" \
@@ -74,7 +74,7 @@
 	    set CWD [pwd]
 	    cd $linkdir
 	    catch "unlink X" ret
-	    if [catch "link $Xwinhome/bin/XF86_$server X" ret] {
+	    if [catch "link ../../usr/X11R6/bin/XF86_$server X" ret] {
 		mesg $messages(phase5.4) okay
 	    } else {
 		mesg $messages(phase5.5) okay
diff -Nur XFree86-4.0.2.org/xc/programs/xdm/config/Xres.cpp XFree86-4.0.2/xc/programs/xdm/config/Xres.cpp
--- XFree86-4.0.2.org/xc/programs/xdm/config/Xres.cpp	Wed Nov 29 17:46:09 2000
+++ XFree86-4.0.2/xc/programs/xdm/config/Xres.cpp	Wed Dec 20 13:41:03 2000
@@ -66,9 +66,9 @@
 /**/#endif
 #ifdef XPM
 /**/#if PLANES >= 8
-xlogin*logoFileName: BITMAPDIR/**//XDM_PIXMAP
+xlogin*logoFileName: BITMAPDIR/**//xdm-pld-logo.xpm
 /**/#else
-xlogin*logoFileName: BITMAPDIR/**//XDM_BWPIXMAP
+xlogin*logoFileName: BITMAPDIR/**//xdm-pld-logo-bw.xpm
 /**/#endif
 xlogin*useShape: true
 xlogin*logoPadding: 10
diff -Nur XFree86-4.0.2.org/xc/programs/xdm/resource.c XFree86-4.0.2/xc/programs/xdm/resource.c
--- XFree86-4.0.2.org/xc/programs/xdm/resource.c	Sun Dec  6 07:08:49 1998
+++ XFree86-4.0.2/xc/programs/xdm/resource.c	Wed Dec 20 13:41:03 2000
@@ -70,31 +70,31 @@
  */
 #ifndef __EMX__
 #ifndef DEF_SERVER_LINE 
-#define DEF_SERVER_LINE ":0 local /usr/bin/X11/X :0"
+#define DEF_SERVER_LINE ":0 local /usr/X11R6/bin/X :0"
 #endif
 #ifndef XRDB_PROGRAM
-#define XRDB_PROGRAM "/usr/bin/X11/xrdb"
+#define XRDB_PROGRAM "/usr/X11R6/X11/xrdb"
 #endif
 #ifndef DEF_SESSION
-#define DEF_SESSION "/usr/bin/X11/xterm -ls"
+#define DEF_SESSION "/usr/X11R6/bin/xterm -ls"
 #endif
 #ifndef DEF_USER_PATH
-#define DEF_USER_PATH ":/bin:/usr/bin:/usr/bin/X11:/usr/ucb"
+#define DEF_USER_PATH "/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin"
 #endif
 #ifndef DEF_SYSTEM_PATH
-#define DEF_SYSTEM_PATH "/etc:/bin:/usr/bin:/usr/bin/X11:/usr/ucb"
+#define DEF_SYSTEM_PATH "/sbin:/bin:/usr/sbin:/usr/bin:/usr/X11R6/bin"
 #endif
 #ifndef DEF_SYSTEM_SHELL
 #define DEF_SYSTEM_SHELL "/bin/sh"
 #endif
 #ifndef DEF_FAILSAFE_CLIENT
-#define DEF_FAILSAFE_CLIENT "/usr/bin/X11/xterm"
+#define DEF_FAILSAFE_CLIENT "/usr/X11R6/bin/xterm"
 #endif
 #ifndef DEF_XDM_CONFIG
-#define DEF_XDM_CONFIG "/usr/lib/X11/xdm/xdm-config"
+#define DEF_XDM_CONFIG "/etc/X11/xdm/xdm-config"
 #endif
 #ifndef DEF_CHOOSER
-#define DEF_CHOOSER "/usr/lib/X11/xdm/chooser"
+#define DEF_CHOOSER "/etc/X11/xdm/chooser"
 #endif
 #ifndef DEF_AUTH_NAME
 #ifdef HASXDMAUTH
@@ -104,7 +104,7 @@
 #endif
 #endif
 #ifndef DEF_AUTH_DIR
-#define DEF_AUTH_DIR "/usr/lib/X11/xdm"
+#define DEF_AUTH_DIR "/etc/X11/xdm"
 #endif
 #ifndef DEF_USER_AUTH_DIR
 #define DEF_USER_AUTH_DIR	"/tmp"
diff -Nur XFree86-4.0.2.org/xc/programs/xinit/Imakefile XFree86-4.0.2/xc/programs/xinit/Imakefile
--- XFree86-4.0.2.org/xc/programs/xinit/Imakefile	Tue Dec  5 01:59:44 2000
+++ XFree86-4.0.2/xc/programs/xinit/Imakefile	Wed Dec 20 13:41:03 2000
@@ -41,11 +41,13 @@
 InstallNamedProg(startx,startx,$(BINDIR))
 InstallManPage(startx,$(MANDIR))
 
-#if InstallXinitConfig
-InstallNonExecFile($(SAMPLECONFIG),$(XINITDIR))
-#else
-InstallNonExecFileNoClobber($(SAMPLECONFIG),$(XINITDIR))
-#endif
+XCOMM Red Hat gets this from a different package
+XCOMM #if InstallXinitConfig
+XCOMM InstallNonExecFile($(SAMPLECONFIG),$(XINITDIR))
+XCOMM #else
+XCOMM InstallNonExecFileNoClobber($(SAMPLECONFIG),$(XINITDIR))
+XCOMM #endif
+
 #endif
 
 LinkConfDirectory(xinit,.,xinit,.)
diff -Nur XFree86-4.0.2.org/xc/programs/xterm/Imakefile XFree86-4.0.2/xc/programs/xterm/Imakefile
--- XFree86-4.0.2.org/xc/programs/xterm/Imakefile	Tue Dec  5 02:00:14 2000
+++ XFree86-4.0.2/xc/programs/xterm/Imakefile	Wed Dec 20 13:43:30 2000
@@ -36,6 +36,10 @@
 #define XkbClientLibs /**/
 #endif
 
+#if InstallXtermSetUID
+#undef InstallXtermSetUID
+#endif
+
 #ifndef SpecialCObjectRule
 #define SpecialCObjectRule(module,ignore,defines) \
 module.o: ; $(CC) -c defines $(CFLAGS) module.c
@@ -57,6 +61,8 @@
          PTYLIB = -lpucc
 #endif
 
+         PTYLIB = -lutil
+
 #if ((LinuxCLibMajorVersion == 6) && (LinuxCLibMinorVersion < 1))
          PTYLIB = -lutil
 #endif
@@ -109,7 +115,7 @@
 		  -DOSMINORVERSION=$(OSMINORVERSION)
    MISC_DEFINES = /* -DALLOWLOGGING -DALLOWLOGFILEEXEC */
     XKB_DEFINES = XkbClientDefines
-        DEFINES = -I. $(XKB_DEFINES) $(TERMCAPDEFINES) $(FEATURE_DEFINES) $(SCROLLBAR_RIGHT) $(UTF8_OPTION) $(XRFDEF)
+        DEFINES = -I. -I/usr/include/ncurses $(XKB_DEFINES) $(TERMCAPDEFINES) $(FEATURE_DEFINES) $(SCROLLBAR_RIGHT) $(UTF8_OPTION) $(XRFDEF)
 
 #ifdef OS2Architecture
           MAINSRC = os2main.c
diff -Nur XFree86-4.0.2.org/xc/programs/xterm/xterm.h XFree86-4.0.2/xc/programs/xterm/xterm.h
--- XFree86-4.0.2.org/xc/programs/xterm/xterm.h	Sat Dec  9 05:02:44 2000
+++ XFree86-4.0.2/xc/programs/xterm/xterm.h	Wed Dec 20 13:41:03 2000
@@ -62,7 +62,7 @@
 #endif
 
 #ifndef DFT_TERMTYPE
-#define DFT_TERMTYPE "xterm"
+#define DFT_TERMTYPE "xterm-color"
 #endif
 
 #ifndef X_NOT_POSIX
--- XFree86-4.0.2/xc/programs/Xserver/hw/xfree86/xf86cfg/Imakefile~	Thu Dec 14 20:47:46 2000
+++ XFree86-4.0.2/xc/programs/Xserver/hw/xfree86/xf86cfg/Imakefile	Fri Dec 22 17:58:27 2000
@@ -92,7 +92,7 @@
 
 #if HasNCurses
 CURSESLIB = NCursesLibName
-CURSESDEFINES  = -DHAS_NCURSES
+CURSESDEFINES  = -DHAS_NCURSES -I/usr/include/ncurses
 #endif
 
 DEFINES        = -DXF86CONFIG=\"XConfigFile\" $(MODULEDEFINES) \
