diff -urN ati.old.org/ati.c ati.old/ati.c
--- ati.old.org/ati.c	Wed Mar 13 11:18:18 2002
+++ ati.old/ati.c	Wed Mar 13 21:05:26 2002
@@ -90,7 +90,7 @@
 DriverRec ATI =
 {
     ATI_VERSION_CURRENT,
-    "ati",
+    "ati_old",
     ATIIdentify,
     ATIProbe,
     ATIAvailableOptions,
diff -urN ati.old.org/atimisc.c ati.old/atimisc.c
--- ati.old.org/atimisc.c	Wed Mar 13 11:18:18 2002
+++ ati.old/atimisc.c	Wed Mar 13 11:19:27 2002
@@ -71,7 +71,7 @@
 }
 
 /* The following record must be called atimiscModuleData */
-XF86ModuleData atimiscModuleData =
+XF86ModuleData atimisc_oldModuleData =
 {
     &ATIVersionRec,
     ATISetup,
diff -urN ati.old.org/atimodule.c ati.old/atimodule.c
--- ati.old.org/atimodule.c	Wed Mar 13 11:18:18 2002
+++ ati.old/atimodule.c	Wed Mar 13 11:19:40 2002
@@ -116,7 +116,7 @@
 }
 
 /* The following record must be called atiModuleData */
-XF86ModuleData atiModuleData =
+XF86ModuleData ati_oldModuleData =
 {
     &ATIVersionRec,
     ATISetup,
diff -urN ati.old.org/atiprobe.c ati.old/atiprobe.c
--- ati.old.org/atiprobe.c	Wed Mar 13 11:18:44 2002
+++ ati.old/atiprobe.c	Mon Mar 11 13:14:51 2002
@@ -2275,10 +2275,10 @@
 
 #ifdef XFree86LOADER
 
-            if (!xf86LoadSubModule(pScreenInfo, "atimisc"))
+            if (!xf86LoadSubModule(pScreenInfo, "atimisc_old"))
             {
                 xf86Msg(X_ERROR,
-                    ATI_NAME ":  Failed to load \"atimisc\" module.\n");
+                    ATI_NAME ":  Failed to load \"atimisc_old\" module.\n");
                 xf86DeleteScreen(pScreenInfo->scrnIndex, 0);
                 continue;
             }
diff -urN ati.old.org/ativersion.h ati.old/ativersion.h
--- ati.old.org/ativersion.h	Wed Mar 13 11:18:44 2002
+++ ati.old/ativersion.h	Mon Mar 11 13:14:51 2002
@@ -25,7 +25,7 @@
 #define ___ATIVERSION_H___ 1
 
 #define ATI_NAME          "ATI"
-#define ATI_DRIVER_NAME   "ati"
+#define ATI_DRIVER_NAME   "ati_old"
 
 #define ATI_VERSION_NAME  "6.4.8"
 
diff -urN ati.old.org/r128_misc.c ati.old/r128_misc.c
--- ati.old.org/r128_misc.c	Wed Mar 13 11:18:18 2002
+++ ati.old/r128_misc.c	Wed Mar 13 11:20:02 2002
@@ -74,7 +74,7 @@
 }
 
 /* The following record must be called r128ModuleData */
-XF86ModuleData r128ModuleData =
+XF86ModuleData r128_oldModuleData =
 {
     &R128VersionRec,
     R128Setup,
diff -urN ati.old.org/r128_probe.c ati.old/r128_probe.c
--- ati.old.org/r128_probe.c	Wed Mar 13 11:18:44 2002
+++ ati.old/r128_probe.c	Mon Mar 11 13:14:51 2002
@@ -205,9 +205,9 @@
 
 #ifdef XFree86LOADER
 
-	    if (!xf86LoadSubModule(pScrn, "r128")) {
+	    if (!xf86LoadSubModule(pScrn, "r128_old")) {
 		xf86Msg(X_ERROR,
-		    R128_NAME ":  Failed to load \"r128\" module.\n");
+		    R128_NAME ":  Failed to load \"r128_old\" module.\n");
 		xf86DeleteScreen(pScrn->scrnIndex, 0);
 		continue;
 	    }
diff -urN ati.old.org/r128_version.h ati.old/r128_version.h
--- ati.old.org/r128_version.h	Wed Mar 13 11:18:18 2002
+++ ati.old/r128_version.h	Wed Mar 13 21:07:17 2002
@@ -25,7 +25,7 @@
 #define _R128_VERSION_H_ 1
 
 #define R128_NAME          "R128"
-#define R128_DRIVER_NAME   "r128"
+#define R128_DRIVER_NAME   "r128_old"
 
 #define R128_VERSION_NAME  "4.0.1"
 
diff -urN ati.old.org/radeon_misc.c ati.old/radeon_misc.c
--- ati.old.org/radeon_misc.c	Wed Mar 13 11:18:18 2002
+++ ati.old/radeon_misc.c	Wed Mar 13 11:20:20 2002
@@ -74,7 +74,7 @@
 }
 
 /* The following record must be called radeonModuleData */
-XF86ModuleData radeonModuleData =
+XF86ModuleData radeon_oldModuleData =
 {
     &RADEONVersionRec,
     RADEONSetup,
diff -urN ati.old.org/radeon_probe.c ati.old/radeon_probe.c
--- ati.old.org/radeon_probe.c	Wed Mar 13 11:18:44 2002
+++ ati.old/radeon_probe.c	Mon Mar 11 13:14:51 2002
@@ -203,9 +203,9 @@
         { 
 #ifdef XFree86LOADER
 
-	    if (!xf86LoadSubModule(pScrn, "radeon")) {
+	    if (!xf86LoadSubModule(pScrn, "radeon_old")) {
 		xf86Msg(X_ERROR,
-		    RADEON_NAME ":  Failed to load \"radeon\" module.\n");
+		    RADEON_NAME ":  Failed to load \"radeon_old\" module.\n");
 		xf86DeleteScreen(pScrn->scrnIndex, 0);
 		continue;
 	    }
diff -urN ati.old.org/radeon_version.h ati.old/radeon_version.h
--- ati.old.org/radeon_version.h	Wed Mar 13 11:18:18 2002
+++ ati.old/radeon_version.h	Wed Mar 13 21:07:52 2002
@@ -25,7 +25,7 @@
 #define _RADEON_VERSION_H_ 1
 
 #define RADEON_NAME          "RADEON"
-#define RADEON_DRIVER_NAME   "radeon"
+#define RADEON_DRIVER_NAME   "radeon_old"
 
 #define RADEON_VERSION_NAME  "4.0.1"
 
