--- xc/Makefile	2006-06-11 00:28:00.305505872 +0200
+++ xc/Makefile-new	2006-06-11 00:31:06.181248464 +0200
@@ -3,7 +3,7 @@
 # Luna users will need to either run make as "make MAKE=make"
 # or add "MAKE = make" to this file.
 
-SHELL = /bin/sh
+SHELL = /bin/sh -e
 RM = rm -f
 MV = mv
 WORLDOPTS =
--- xc/config/cf/Imake.cf.orig	2006-01-09 15:56:12.000000000 +0100
+++ xc/config/cf/Imake.cf	2006-06-11 00:28:00.321503440 +0200
@@ -800,6 +800,10 @@
 #   define s390Architecture
 #  undef __s390__
 # endif /* s390 */
+# ifdef alpha
+#  define AlphaArchitecture
+#  undef alpha
+# endif /* alpha */
 # ifdef __alpha
 #  define AlphaArchitecture
 #  undef __alpha
--- xc/config/cf/Imake.tmpl.orig	2006-04-03 01:12:46.000000000 +0200
+++ xc/config/cf/Imake.tmpl	2006-06-11 00:28:00.323503136 +0200
@@ -308,7 +308,7 @@
 
 #ifndef BuildHtmlManPages
 #ifdef RmanCmd
-#define BuildHtmlManPages YES
+#define BuildHtmlManPages NO
 #else
 #define BuildHtmlManPages NO
 #endif
@@ -904,11 +904,7 @@
 #define LintlibDir $(USRLIBDIR)/lint
 #endif
 #ifndef SystemManDirectory
-#if SystemV4
 #define SystemManDirectory /usr/share/man
-#else
-#define SystemManDirectory /usr/man
-#endif
 #endif
 #ifndef ManDirectoryRoot
 #ifdef ProjectRoot
@@ -1055,7 +1051,7 @@
 #define NoRConst NO /* YES if const for structs of funcs is bad */
 #endif
 #ifndef InstPgmFlags
-#define InstPgmFlags -s
+#define InstPgmFlags
 #endif
 #ifndef InstBinFlags
 #define InstBinFlags -m 0755
@@ -1212,7 +1208,7 @@
 #define CppCmd /LibDirName/cpp
 #endif
 #ifndef RawCppCmd
-#define RawCppCmd CppCmd -undef
+#define RawCppCmd CppCmd
 #endif
 #ifndef CppNoLineInfoOption
 #define CppNoLineInfoOption /**/
--- xc/config/cf/X11.tmpl.orig	2006-04-18 17:57:25.000000000 +0200
+++ xc/config/cf/X11.tmpl	2006-06-11 00:28:00.327502528 +0200
@@ -1649,8 +1649,12 @@
 # define UsbMouseSupport NO
 #endif
 
+#ifndef EtcX11Directory
+#define EtcX11Directory /etc/X11
+#endif
+
 #ifndef DefaultFSConfigFile
-#define DefaultFSConfigFile $(LIBDIR)/fs/config
+#define DefaultFSConfigFile $(ETCX11DIR)/fs/config
 #endif
 
 /* X-TrueType settings */
@@ -1783,7 +1787,7 @@
 #define ConfigDir $(LIBDIR)/config
 #endif
 #ifndef DocDir
-#define DocDir $(LIBDIR)/doc
+#define DocDir $(LIBDIR)/share/doc
 #endif
 #ifndef DocHtmlDir
 #define DocHtmlDir $(DOCDIR)/html
@@ -1795,31 +1799,31 @@
 #define DocPdfDir $(DOCDIR)/PDF
 #endif
 #ifndef FontDir
-#define FontDir $(LIBDIR)/fonts
+#define FontDir /usr/share/fonts
 #endif
 #ifndef AdmDir
-#define AdmDir /usr/adm
+#define AdmDir /var/log
 #endif
 #ifndef XinitDir
-#define XinitDir $(LIBDIR)/xinit
+#define XinitDir $(ETCX11DIR)/xinit
 #endif
 #ifndef XdmDir
-#define XdmDir $(LIBDIR)/xdm
+#define XdmDir $(ETCX11DIR)/xdm
 #endif
 #ifndef XdmVarDir
 #define XdmVarDir $(VARLIBDIR)/xdm
 #endif
 #ifndef XdmAuthDir
-#define XdmAuthDir $(LIBDIR)/xdm
+#define XdmAuthDir $(ETCX11DIR)/xdm
 #endif
 #ifndef XdmConfigurationSubdirectory		/* under xdm sources */
 #define XdmConfigurationSubdirectory config
 #endif
 #ifndef TwmDir
-#define TwmDir $(LIBDIR)/twm
+#define TwmDir $(ETCX11DIR)/twm
 #endif
 #ifndef XsmDir
-#define XsmDir $(LIBDIR)/xsm
+#define XsmDir $(ETCX11DIR)/xsm
 #endif
 #ifndef NlsDir
 #define NlsDir $(LIBDIR)/nls
@@ -1828,23 +1832,20 @@
 #define XLocaleDir $(LIBDIR)/locale
 #endif
 #ifndef LbxproxyDir
-#define LbxproxyDir $(LIBDIR)/lbxproxy
+#define LbxproxyDir $(ETCX11DIR)/lbxproxy
 #endif
 #ifndef LbxproxyConfigSubdir		/* under lbxproxy sources */
 #define LbxproxyConfigSubdir config
 #endif
 #ifndef ProxyManagerDir
-#define ProxyManagerDir $(LIBDIR)/proxymngr
+#define ProxyManagerDir $(ETCX11DIR)/proxymngr
 #endif
 #ifndef ServerConfigDir
-#define ServerConfigDir $(LIBDIR)/xserver
+#define ServerConfigDir $(ETCX11DIR)/xserver
 #endif
 #ifndef XPrintDir
 #define XPrintDir ServerConfigDir
 #endif
-#ifndef EtcX11Directory
-#define EtcX11Directory /etc/X11
-#endif
 #ifndef InstallSecurityConfig
 #define InstallSecurityConfig YES	/* Install sample security policy */
 #endif
@@ -1876,10 +1877,10 @@
 #define OsNameDefines /* as nothing */
 #endif
 #ifndef DefaultUserPath			/* for xdm */
-#define DefaultUserPath :/bin:/usr/bin:$(BINDIR):/usr/ucb
+#define DefaultUserPath /usr/local/bin:/bin:/usr/bin:$(BINDIR)
 #endif
 #ifndef DefaultSystemPath		/* for xdm */
-#define DefaultSystemPath /etc:/bin:/usr/bin:$(BINDIR):/usr/ucb
+#define DefaultSystemPath /sbin:/bin:/usr/sbin:/usr/bin:$(BINDIR)
 #endif
 #ifndef DefaultSystemShell		/* for xdm to set in SHELL */
 #define DefaultSystemShell BourneShell
@@ -2014,7 +2015,7 @@
 
      HTMLINDEXCMD = HtmlIndexCmd
 
-       DOCUTILSRC = $(XTOP)/doc/util
+       DOCUTILSRC = $(XTOP)/share/doc/util
         CLIENTSRC = $(TOP)/clients
           DEMOSRC = $(TOP)/demos
        XDOCMACROS = $(DOCUTILSRC)/macros.t
@@ -2028,7 +2029,7 @@
         SERVERSRC = $(XTOP)/programs/Xserver
        CONTRIBSRC = ContribDir
    UNSUPPORTEDSRC = UnsupportedDir
-           DOCSRC = $(XTOP)/doc
+           DOCSRC = $(XTOP)/share/doc
            RGBSRC = $(XTOP)/programs/rgb
       BDFTOPCFSRC = $(PROGRAMSRC)/bdftopcf
      MKFONTDIRSRC = $(PROGRAMSRC)/mkfontdir
--- xc/config/cf/linux.cf.orig	2006-04-03 01:12:46.000000000 +0200
+++ xc/config/cf/linux.cf	2006-06-11 00:28:00.329502224 +0200
@@ -335,7 +335,7 @@
  */
 #if BuildXF86DRI
 # if defined(i386Architecture)
-#  define BuildLibGlxWithoutPIC	YES
+#  define BuildLibGlxWithoutPIC	NO
 # endif
 #endif
 
@@ -749,7 +749,7 @@
 
 #define HasVarRun		YES
 #ifndef VarDbDirectory
-# define VarDbDirectory		$(VARDIR)/lib
+# define VarDbDirectory		/var/lib
 #endif
 
 #ifndef XF86OSCardDrivers
--- xc/config/cf/site.def.orig	2006-01-09 15:56:13.000000000 +0100
+++ xc/config/cf/site.def	2006-06-11 00:28:00.330502072 +0200
@@ -104,10 +104,10 @@
 #define HasSentinel	YES
 */
 
-/*
 #undef DefaultUserPath
-#define DefaultUserPath /bin:/usr/bin:$(BINDIR):/usr/ucb:/usr/local/bin
-*/
+#define DefaultUserPath /usr/local/bin:/bin:/usr/bin:$(BINDIR)
+#undef DefaultSystemPath
+#define DefaultSystemPath /sbin:/bin:/usr/sbin:/usr/bin:$(BINDIR)
 
 
 /* You do NOT need SetUID if you only run the server under xdm */
--- xc/config/cf/xf86site.def.orig	2006-04-08 20:33:00.000000000 +0200
+++ xc/config/cf/xf86site.def	2006-06-11 00:28:00.335501312 +0200
@@ -257,8 +257,8 @@
 /*
  * To not build/install the Cyrillic fonts, uncomment the following
  *
-#define BuildCyrillicFonts	NO
  */
+#define BuildCyrillicFonts	YES
 
 /*
  * To not install the local font directory, uncomment the following
@@ -304,27 +304,26 @@
  * HasGlide2 to YES and set the Glide2IncDir variable. 
  * HasGlide2 is per default NO.
  *
-#define HasGlide2 YES
  */
+#define HasGlide2 YES
 
 /*
  * Set the path to your Glide include files.
  *
-#define Glide2IncDir /usr/include/glide
  */
+#define Glide2IncDir /usr/include/glide
 
 /* 
  * Have glide 3?
  *
-#define HasGlide3 YES
  */
+#define HasGlide3 YES
 
 /*
  * Set the path to your Glide 3 include files.
  *
-#define Glide3IncDir /usr/include/glide3
  */
-
+#define Glide3IncDir /usr/include/glide3
 
 
 /*
@@ -426,8 +425,8 @@
  * Some Linux releases don't have a libtermcap.  In this case you may need
  * to uncomment the following
  *
-#define TermcapLibrary		-lncurses
  */
+#define TermcapLibrary		-ltinfo
 
 /*
  * Build a server that dynamically loads the modules by setting
@@ -602,8 +601,8 @@
  * To forceably build static libraries in addition to shared libraries,
  * uncomment this.
  *
-#define ForceNormalLib		YES
  */
+#define ForceNormalLib		YES
 
 /*
  * Uncomment this if your default tools (eg, gcc, ld, as, etc) are
@@ -691,4 +691,8 @@
 #define X86EMU_LIBPATH /usr/local/lib
  */ 
 
-
+/* these must be before linux.cf is included */
+#define UsePam		YES
+#define HasPam		YES
+#define HasPamMisc	YES
+#define HasFontconfig	YES
--- xc/config/cf/xfree86.cf.orig	2006-04-26 02:28:16.000000000 +0200
+++ xc/config/cf/xfree86.cf	2006-06-11 00:28:00.337501008 +0200
@@ -148,7 +148,7 @@
 #define XConfigFile		XF86Config
 #endif
 #ifndef XConfigDir
-#define XConfigDir		$(LIBDIR)
+#define XConfigDir		$(ETCX11DIR)
 #endif
 #ifndef XLogFile
 #define XLogFile		XFree86
@@ -788,7 +788,7 @@
 
 # ifndef XF86CardDrivers
 #  define XF86CardDrivers	mga glint nv s3 s3virge sis savage \
-				trident chips tdfx fbdev ati \
+				trident chips fbdev ati \
 				DevelDrivers newport \
 				XF86OSCardDrivers XF86ExtraCardDrivers
 # endif
--- xc/config/imake/imake.c.orig	2006-01-09 15:56:14.000000000 +0100
+++ xc/config/imake/imake.c	2006-06-11 00:28:00.339500704 +0200
@@ -1056,6 +1056,7 @@
   static char* suse = "/etc/SuSE-release";
   static char* redhat = "/etc/redhat-release";
   static char* debian = "/etc/debian_version";
+  static char* pld = "/etc/pld-release";
 
   fprintf (inFile, "%s\n", "#define LinuxUnknown    0");
   fprintf (inFile, "%s\n", "#define LinuxSuSE       1");
@@ -1070,6 +1071,7 @@
   fprintf (inFile, "%s\n", "#define LinuxTurbo      10");
   fprintf (inFile, "%s\n", "#define LinuxWare       11");
   fprintf (inFile, "%s\n", "#define LinuxYggdrasil  12");
+  fprintf (inFile, "%s\n", "#define LinuxPLD        13");
 
 #ifdef CROSSCOMPILE
   if (CrossCompiling) {
@@ -1096,6 +1098,11 @@
      * at the content of /etc/debian_version */
     return;
   }
+  if (lstat (pld, &sb) == 0) {
+    fprintf (inFile, "%s\n", "#define DefaultLinuxDistribution LinuxPLD");
+    fprintf (inFile, "%s\n", "#define DefaultLinuxDistName PLD");
+    return;
+  }
   /* what's the definitive way to tell what any particular distribution is? */
 
   fprintf (inFile, "%s\n", "#define DefaultLinuxDistribution LinuxUnknown");
--- xc/programs/Xserver/hw/xfree86/xf86cfg/Imakefile.orig	2006-03-11 18:36:50.000000000 +0100
+++ xc/programs/Xserver/hw/xfree86/xf86cfg/Imakefile	2006-06-11 00:28:00.340500552 +0200
@@ -108,7 +108,7 @@
 
 #if HasNCurses
 CURSESLIB = NCursesLibName
-CURSESDEFINES  = -DHAS_NCURSES
+CURSESDEFINES  = -DHAS_NCURSES -I/usr/include/ncurses
 #endif
 
 XF86CONFIGFILE = XConfigFile
--- xc/programs/xdm/config/Xres.cpp.orig	2006-01-09 16:01:06.000000000 +0100
+++ xc/programs/xdm/config/Xres.cpp	2006-06-11 00:28:00.341500400 +0200
@@ -65,10 +65,10 @@
 #endif /* XPM */
 XHASHendif
 #ifdef XPM
-XHASHif PLANES >= 8
-xlogin*logoFileName: BITMAPDIR/**//XDM_PIXMAP
+XHASHif PLANES > 8
+xlogin*logoFileName: BITMAPDIR/**//xdm-pld-logo.xpm
 XHASHelse
-xlogin*logoFileName: BITMAPDIR/**//XDM_BWPIXMAP
+xlogin*logoFileName: BITMAPDIR/**//xdm-pld-logo-bw.xpm
 XHASHendif
 xlogin*useShape: true
 xlogin*logoPadding: 10
--- xc/programs/xdm/resource.c.orig	2006-01-09 16:01:04.000000000 +0100
+++ xc/programs/xdm/resource.c	2006-06-11 00:28:00.342500248 +0200
@@ -83,31 +83,31 @@
  */
 #ifndef __EMX__
 #ifndef DEF_SERVER_LINE 
-#define DEF_SERVER_LINE ":0 local /usr/bin/X11/X :0"
+#define DEF_SERVER_LINE ":0 local /usr/X11R6/bin/X :0"
 #endif
 #ifndef XRDB_PROGRAM
-#define XRDB_PROGRAM "/usr/bin/X11/xrdb"
+#define XRDB_PROGRAM "/usr/X11R6/X11/xrdb"
 #endif
 #ifndef DEF_SESSION
-#define DEF_SESSION "/usr/bin/X11/xterm -ls"
+#define DEF_SESSION "/usr/X11R6/bin/xterm -ls"
 #endif
 #ifndef DEF_USER_PATH
-#define DEF_USER_PATH ":/bin:/usr/bin:/usr/bin/X11:/usr/ucb"
+#define DEF_USER_PATH "/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin"
 #endif
 #ifndef DEF_SYSTEM_PATH
-#define DEF_SYSTEM_PATH "/etc:/bin:/usr/bin:/usr/bin/X11:/usr/ucb"
+#define DEF_SYSTEM_PATH "/sbin:/bin:/usr/sbin:/usr/bin:/usr/X11R6/bin"
 #endif
 #ifndef DEF_SYSTEM_SHELL
 #define DEF_SYSTEM_SHELL "/bin/sh"
 #endif
 #ifndef DEF_FAILSAFE_CLIENT
-#define DEF_FAILSAFE_CLIENT "/usr/bin/X11/xterm"
+#define DEF_FAILSAFE_CLIENT "/usr/X11R6/bin/xterm"
 #endif
 #ifndef DEF_XDM_CONFIG
-#define DEF_XDM_CONFIG "/usr/lib/X11/xdm/xdm-config"
+#define DEF_XDM_CONFIG "/etc/X11/xdm/xdm-config"
 #endif
 #ifndef DEF_CHOOSER
-#define DEF_CHOOSER "/usr/lib/X11/xdm/chooser"
+#define DEF_CHOOSER "/etc/X11/xdm/chooser"
 #endif
 #ifndef DEF_AUTH_NAME
 #ifdef HASXDMAUTH
@@ -117,7 +117,7 @@
 #endif
 #endif
 #ifndef DEF_AUTH_DIR
-#define DEF_AUTH_DIR "/usr/lib/X11/xdm"
+#define DEF_AUTH_DIR "/etc/X11/xdm"
 #endif
 #ifndef DEF_USER_AUTH_DIR
 #define DEF_USER_AUTH_DIR	"/tmp"
--- xc/programs/xinit/Imakefile.orig	2006-03-06 17:06:23.000000000 +0100
+++ xc/programs/xinit/Imakefile	2006-06-11 00:28:00.342500248 +0200
@@ -41,11 +41,12 @@
 #endif
 InstallManPage(startx,$(MANDIR))
 
-#if InstallXinitConfig
-InstallNonExecFile($(SAMPLECONFIG),$(XINITDIR))
-#else
-InstallNonExecFileNoClobber($(SAMPLECONFIG),$(XINITDIR))
-#endif
+XCOMM PLD gets this from a different package
+XCOMM #if InstallXinitConfig
+XCOMM InstallNonExecFile($(SAMPLECONFIG),$(XINITDIR))
+XCOMM #else
+XCOMM InstallNonExecFileNoClobber($(SAMPLECONFIG),$(XINITDIR))
+XCOMM #endif
 
 LinkConfDirectory(xinit,.,xinit,.)
 
--- xc/programs/xterm/Imakefile.orig	2007-07-03 17:19:53.000000000 +0200
+++ xc/programs/xterm/Imakefile	2007-12-02 16:10:40.453362442 +0100
@@ -102,6 +102,10 @@
 # endif
 #endif
 
+#if InstallXtermSetUID
+#undef InstallXtermSetUID
+#endif
+
 #ifndef SpecialCObjectRule
 #define SpecialCObjectRule(module,ignore,defines) \
 module.o: ; $(CC) -c defines $(CFLAGS) module.c
@@ -204,7 +208,7 @@
         DEFINES = $(XKB_DEFINES) $(TERMCAPDEFINES) $(FEATURE_DEFINES) \
                   $(SCROLLBAR_RIGHT) $(UTF8_OPTION) $(XRFDEF) $(PATH_DEFINES) \
                   $(PUTENVDEF) $(IMAKEDEFINES) $(TRACEDEF)
-       INCLUDES = $(OBSOLETE_INCLUDES) $(XRFINCLUDES)
+       INCLUDES = $(OBSOLETE_INCLUDES) $(XRFINCLUDES) -I/usr/include/ncurses
 
 #ifdef OS2Architecture
         MAINSRC = os2main.c
