--- XFree86-4.2.0/xc/programs/Xserver/hw/xfree86/common/xf86PciInfo.h.orig	Thu Feb 21 14:41:45 2002
+++ XFree86-4.2.0/xc/programs/Xserver/hw/xfree86/common/xf86PciInfo.h	Thu Feb 21 14:56:44 2002
@@ -625,6 +625,8 @@
 #define PCI_CHIP_SAVAGE_MX	0x8c11
 #define PCI_CHIP_SAVAGE_IX_MV	0x8c12
 #define PCI_CHIP_SAVAGE_IX	0x8c13
+#define PCI_CHIP_S3TWISTER_P	0x8d01
+#define PCI_CHIP_S3TWISTER_K	0x8d02
 
 /* ARK Logic */
 #define PCI_CHIP_1000PV		0xA091
@@ -1555,6 +1557,8 @@
 				{PCI_CHIP_SAVAGE_IX_MV,	"Savage/IX-MV",0},
 				{PCI_CHIP_PROSAVAGE_PM,	"ProSavage PM133",0},
 				{PCI_CHIP_PROSAVAGE_KM,	"ProSavage KM133",0},
+				{PCI_CHIP_S3TWISTER_P,	"ProSavage PN133",0},
+				{PCI_CHIP_S3TWISTER_K,	"ProSavage KN133",0},
 				{PCI_CHIP_VIRGE_MX,	"ViRGE/MX",0},
 				{PCI_CHIP_VIRGE_MXPLUS,	"ViRGE/MX+",0},
 				{PCI_CHIP_VIRGE_MXP,	"ViRGE/MX+MV",0},
