diff -urN XFree86-4.0.orig/xc/config/cf/Imake.tmpl XFree86-4.0/xc/config/cf/Imake.tmpl
--- XFree86-4.0.orig/xc/config/cf/Imake.tmpl	Tue Mar  7 16:54:17 2000
+++ XFree86-4.0/xc/config/cf/Imake.tmpl	Fri Mar 17 02:19:49 2000
@@ -555,9 +555,9 @@
 #endif
 #ifndef VarLibDir
 #ifdef ProjectVar
-#define VarLibDir Concat(ProjectVar,/lib)
+#define VarLibDir Concat(ProjectVar,/state)
 #else
-#define VarLibDir /var/lib
+#define VarLibDir /var/state
 #endif
 #endif
 
@@ -585,11 +585,7 @@
 #define LintlibDir $(USRLIBDIR)/lint
 #endif
 #ifndef SystemManDirectory
-#if SystemV4
 #define SystemManDirectory /usr/share/man
-#else
-#define SystemManDirectory /usr/man
-#endif
 #endif
 #ifndef ManDirectoryRoot
 #ifdef ProjectRoot
@@ -710,7 +706,7 @@
 #define NoRConst NO /* YES if const for structs of funcs is bad */
 #endif
 #ifndef InstPgmFlags
-#define InstPgmFlags -s
+#define InstPgmFlags
 #endif
 #ifndef InstBinFlags
 #define InstBinFlags -m 0755
diff -urN XFree86-4.0.orig/xc/config/cf/X11.tmpl XFree86-4.0/xc/config/cf/X11.tmpl
--- XFree86-4.0.orig/xc/config/cf/X11.tmpl	Sat Mar  4 05:35:22 2000
+++ XFree86-4.0/xc/config/cf/X11.tmpl	Fri Mar 17 02:19:49 2000
@@ -946,7 +946,7 @@
 #define ConfigDir $(LIBDIR)/config
 #endif
 #ifndef DocDir
-#define DocDir $(LIBDIR)/doc
+#define DocDir $(LIBDIR)/share/doc
 #endif
 #ifndef DocHtmlDir
 #define DocHtmlDir $(DOCDIR)/html
@@ -955,7 +955,7 @@
 #define DocPsDir $(DOCDIR)/PostScript
 #endif
 #ifndef FontDir
-#define FontDir $(LIBDIR)/fonts
+#define FontDir /usr/share/fonts
 #endif
 #ifndef AdmDir
 #define AdmDir /usr/adm
@@ -1147,7 +1147,7 @@
         MKFONTDIR = $(CLIENTENVSETUP) $(XBUILDBINDIR)/mkfontdir
 #endif
 
-       DOCUTILSRC = $(XTOP)/doc/util
+       DOCUTILSRC = $(XTOP)/share/doc/util
         CLIENTSRC = $(TOP)/clients
           DEMOSRC = $(TOP)/demos
        XDOCMACROS = $(DOCUTILSRC)/macros.t
@@ -1161,7 +1161,7 @@
         SERVERSRC = $(XTOP)/programs/Xserver
        CONTRIBSRC = ContribDir
    UNSUPPORTEDSRC = UnsupportedDir
-           DOCSRC = $(XTOP)/doc
+           DOCSRC = $(XTOP)/share/doc
            RGBSRC = $(XTOP)/programs/rgb
       BDFTOPCFSRC = $(PROGRAMSRC)/bdftopcf
      MKFONTDIRSRC = $(PROGRAMSRC)/mkfontdir
diff -urN XFree86-4.0.orig/xc/config/cf/host.def XFree86-4.0/xc/config/cf/host.def
--- XFree86-4.0.orig/xc/config/cf/host.def	Thu Jan  1 01:00:00 1970
+++ XFree86-4.0/xc/config/cf/host.def	Fri Mar 17 02:19:49 2000
@@ -0,0 +1 @@
+
diff -urN XFree86-4.0.orig/xc/config/cf/linux.cf XFree86-4.0/xc/config/cf/linux.cf
--- XFree86-4.0.orig/xc/config/cf/linux.cf	Sat Mar  4 05:35:22 2000
+++ XFree86-4.0/xc/config/cf/linux.cf	Fri Mar 17 02:19:49 2000
@@ -192,7 +192,7 @@
 #  ifdef i386Architecture
 #    if OSMajorVersion == 2
        /* Only tested for Linux 2.2.0 through 2.2.12 */
-#      if OSMinorVersion == 2 && OSTeenyVersion <= 12
+#      if OSMinorVersion == 2
 #        define BuildXF86DRM        YES
 #      else
          /* Only tested for Linux 2.3.0 through 2.3.18 */
diff -urN XFree86-4.0.orig/xc/config/cf/site.def XFree86-4.0/xc/config/cf/site.def
--- XFree86-4.0.orig/xc/config/cf/site.def	Mon Oct  5 05:52:26 1998
+++ XFree86-4.0/xc/config/cf/site.def	Fri Mar 17 02:19:49 2000
@@ -115,10 +115,10 @@
 #define HasSentinel	YES
 */
 
-/*
 #undef DefaultUserPath
-#define DefaultUserPath /bin:/usr/bin:$(BINDIR):/usr/ucb:/usr/local/bin
-*/
+#define DefaultUserPath /bin:/usr/bin:/usr/local/bin
+#undef DefaultSystemPath
+#define DefaultSystemPath /sbin:/usr/sbin:/bin:/usr/bin
 
 
 /* You do NOT need SetUID if you only run the server under xdm */
diff -urN XFree86-4.0.orig/xc/config/cf/xf86site.def XFree86-4.0/xc/config/cf/xf86site.def
--- XFree86-4.0.orig/xc/config/cf/xf86site.def	Tue Mar  7 16:54:18 2000
+++ XFree86-4.0/xc/config/cf/xf86site.def	Fri Mar 17 02:21:18 2000
@@ -134,16 +134,16 @@
  * Some Linux releases don't have a libtermcap.  In this case you may need
  * to uncomment the following
  *
-#define TermcapLibrary		-lncurses
  */
+#define TermcapLibrary		-lncurses
 
 /*
  * If you have Tk (which is required to build XF86Setup), uncomment this
  * Note: version 4.0 or 4.1 is required, and XF86Setup links it statically by
  * default.
  *
-#define HasTk			YES
  */
+#define HasTk			YES
 
 /*
  * Set the paths and names for your Tk library if they don't match the
@@ -156,25 +156,25 @@
  *
  * Some Linux releases need TkLibDir set to /usr/X11R6/lib
  *
-#define TkLibDir		/usr/local/lib
-#define TkIncDir		/usr/local/include
-#define TkLibName		tk
  */
+#define TkLibDir		/usr/lib
+#define TkIncDir		/usr/include
+#define TkLibName		tk
 
 /*
  * By default, XF86Setup is linked with a static Tk library.  To change
  * that, uncomment the following line.
  *
-#define XF86SetupUsesStaticTk	NO
  */
+#define XF86SetupUsesStaticTk	NO
 
 /*
  * If you have Tcl (which is required to build XF86Setup), uncomment this
  * Note: version 7.4 or 7.5 is required, and XF86Setup links it statically by
  * default.
  *
-#define HasTcl			YES
  */
+#define HasTcl			YES
 
 /*
  * Set the paths and names for your Tcl library if they don't match the
@@ -184,17 +184,17 @@
  *
  * Some Linux releases need TclIncDir set to /usr/include/tcl.
  *
-#define TclLibDir		/usr/local/lib
-#define TclIncDir		/usr/local/include
-#define TclLibName		tcl
  */
+#define TclLibDir		/usr/lib
+#define TclIncDir		/usr/include
+#define TclLibName		tcl
 
 /*
  * By default, XF86Setup is linked with a static Tcl library.  To change
  * that, uncomment the following line.
  *
-#define XF86SetupUsesStaticTcl	NO
  */
+#define XF86SetupUsesStaticTcl	NO
 
 /*
  * Build a server that dynamically loads the modules by setting
@@ -378,8 +378,8 @@
 /*
  * To not build/install the Cyrillic fonts, uncomment the following
  *
-#define BuildCyrillicFonts	NO
  */
+#define BuildCyrillicFonts	YES
 
 /*
  * To not build/install the Latin2 fonts, uncomment the following
@@ -568,8 +568,8 @@
  * To forceably build static libraries in addition to shared libraries,
  * uncomment this.
  *
-#define ForceNormalLib		YES
  */
+#define ForceNormalLib		YES
 
 /*
  * Uncomment this if your default tools (eg, gcc, ld, as, etc) are
@@ -655,26 +655,26 @@
  * HasGlide2 to YES and set the Glide2IncDir variable. 
  * HasGlide2 is per default NO.
  *
-#define HasGlide2 YES
  */
+#define HasGlide2 YES
 
 /*
  * Set the path to your Glide include files.
  *
-#define Glide2IncDir /usr/include/glide
  */
+#define Glide2IncDir /usr/include/glide
 
 /* 
  * The 3Dfx 3D DRI driver requires glide 3.
  *
-#define HasGlide3 YES
  */
+#define HasGlide3 YES
 
 /*
  * Set the path to your Glide 3 include files.
  *
-#define Glide3IncDir /usr/include/glide3
  */
+#define Glide3IncDir /usr/include/glide3
 
 /*
  * If your system doesn't support vm86() mode and you have
diff -urN XFree86-4.0.orig/xc/config/imake/Makefile.proto XFree86-4.0/xc/config/imake/Makefile.proto
--- XFree86-4.0.orig/xc/config/imake/Makefile.proto	Thu Jan  1 01:00:00 1970
+++ XFree86-4.0/xc/config/imake/Makefile.proto	Fri Mar 17 02:19:49 2000
@@ -0,0 +1,860 @@
+# Makefile generated by imake - do not edit!
+# $TOG: imake.c /main/104 1998/03/24 12:45:15 kaleb $
+
+# ----------------------------------------------------------------------
+# Makefile generated from "Imake.tmpl" and <./config/imake/Imakefile>
+# $TOG: Imake.tmpl /main/249 1997/10/13 15:28:56 kaleb $
+#
+#
+#
+#
+# $XFree86: xc/config/cf/Imake.tmpl,v 3.65 2000/01/31 19:33:18 dawes Exp $
+# ----------------------------------------------------------------------
+
+all::
+
+.SUFFIXES: .i
+
+# $TOG: Imake.cf /main/30 1998/04/28 13:55:25 barstow $
+
+# $XFree86: xc/config/cf/Imake.cf,v 3.53 2000/02/12 20:45:01 dawes Exp $
+
+# -----------------------------------------------------------------------
+# site-specific configuration parameters that need to come before
+# the platform-specific parameters - edit site.def to change
+
+# site:  $TOG: site.sample /main/r64_final/1 1998/02/05 16:28:49 kaleb $
+
+# site:  $XFree86: xc/config/cf/site.def,v 3.22 1998/10/02 06:15:06 dawes Exp $
+
+# $XFree86: xc/config/cf/xf86site.def,v 3.153 1999/12/03 19:17:17 eich Exp $
+
+# ----------------------------------------------------------------------
+# platform-specific configuration parameters - edit linux.cf to change
+
+# platform:  $TOG: linux.cf /main/47 1998/04/17 11:32:51 mgreess
+# platform:  $XFree86: xc/config/cf/linux.cf,v 3.115 2000/02/21 18:05:31 dawes Exp $
+
+# operating system:  Linux 2.2.14 i586 [ELF] (2.2.14)
+# libc:	(6.1.2)
+# binutils:	(29)
+
+# $TOG: lnxLib.rules /main/14 1998/03/16 11:15:30 mgreess $
+# $XFree86: xc/config/cf/lnxLib.rules,v 3.37 1999/06/06 08:47:44 dawes Exp $
+
+# $XFree86: xc/config/cf/xfree86.cf,v 3.269 2000/02/17 13:50:02 dawes Exp $
+
+# $TOG: xfree86.cf /main/38 1997/10/22 13:01:59 kaleb $
+
+AFB_DEFS = -DUSE_AFB
+
+DRIVERSDKDIR = $(USRLIBDIR)/Server
+DRIVERSDKMODULEDIR = $(USRLIBDIR)/Server/modules
+DRIVERSDKINCLUDEDIR = $(USRLIBDIR)/Server/include
+
+       XF86SRC = $(SERVERSRC)/hw/xfree86
+  XF86ACCELSRC = $(XF86SRC)/accel
+    XF86COMSRC = $(XF86SRC)/common
+ XF86PARSERSRC = $(XF86SRC)/parser
+     XF86OSSRC = $(XF86SRC)/os-support
+ XF86DRIVERSRC = $(XF86SRC)/drivers
+     DRIVERSRC = $(XF86DRIVERSRC)
+   S3DRIVERSRC = $(XF86SRC)/accel/s3/drivers
+  S3VDRIVERSRC = $(XF86SRC)/accel/s3_virge/drivers
+
+       XF68SRC = $(SERVERSRC)/hw/xfree68
+    XF68COMSRC = $(XF68SRC)/common
+ XF68PARSERSRC = $(XF68SRC)/parser
+     XF68OSSRC = $(XF68SRC)/os-support
+
+           XF98SRC = $(SERVERSRC)/hw/xfree98
+      XF98ACCELSRC = $(XF98SRC)/accel
+        XF98COMSRC = $(XF98SRC)/common
+         XF98OSSRC = $(XF98SRC)/os-support
+     XF98DRIVERSRC = $(XF98SRC)/drivers
+  XF98VGADRIVERSRC = $(XF98SRC)/vga256/drivers
+XF98VGA16DRIVERSRC = $(XF98SRC)/vga16/drivers
+ XF98VGA2DRIVERSRC = $(XF98SRC)/vga2/drivers
+ XF98MONODRIVERSRC = $(XF98SRC)/mono/drivers
+XF98NECS3DRIVERSRC = $(XF98SRC)/accel/s3nec/drivers
+XF98PWSKBDRIVERSRC = $(XF98SRC)/accel/s3pwskb/drivers
+ XF98PWLBDRIVERSRC = $(XF98SRC)/accel/s3pwlb/drivers
+XF98GA968DRIVERSRC = $(XF98SRC)/accel/s3ga968/drivers
+
+        XFREE86DOCDIR = $(DOCDIR)
+      XFREE86PSDOCDIR = $(DOCPSDIR)
+    XFREE86HTMLDOCDIR = $(DOCHTMLDIR)
+XFREE86JAPANESEDOCDIR = $(DOCDIR)/Japanese
+
+# $XConsortium: xf86.rules /main/9 1996/10/31 14:54:26 kaleb $
+# $XFree86: xc/config/cf/xf86.rules,v 3.30 1999/10/13 16:49:02 dawes Exp $
+
+# ----------------------------------------------------------------------
+# site-specific configuration parameters that go after
+# the platform-specific parameters - edit site.def to change
+
+# site:  $TOG: site.sample /main/r64_final/1 1998/02/05 16:28:49 kaleb $
+
+# site:  $XFree86: xc/config/cf/site.def,v 3.22 1998/10/02 06:15:06 dawes Exp $
+
+# ---------------------------------------------------------------------
+# Imake rules for building libraries, programs, scripts, and data files
+# rules:  $TOG: Imake.rules /main/227 1998/02/02 12:07:57 kaleb $
+#
+#
+#
+#
+# rules:  $XFree86: xc/config/cf/Imake.rules,v 3.64 2000/02/13 06:15:16 dawes Exp $
+
+ _NULLCMD_ = @ echo -n
+
+GLIDEINCDIR =
+
+TKLIBNAME =
+
+TKLIBDIR =
+
+TCLLIBNAME =
+
+TCLIBDIR =
+
+          PATHSEP = /
+            SHELL = /bin/sh
+
+              TOP = ../..
+      CURRENT_DIR = ./config/imake
+
+            IMAKE = $(IMAKESRC)/imake
+           DEPEND = $(DEPENDSRC)/makedepend
+        MKDIRHIER = mkdir -p
+          REVPATH = $(CONFIGSRC)/util/revpath
+    EXPORTLISTGEN =
+        CONFIGSRC = $(TOP)/config
+         IMAKESRC = $(CONFIGSRC)/imake
+        DEPENDSRC = $(CONFIGSRC)/makedepend
+
+          INCROOT = /usr/X11R6/include
+        USRLIBDIR = /usr/X11R6/lib
+        VARLIBDIR = /var/lib
+         SHLIBDIR = /usr/X11R6/lib
+       LINTLIBDIR = $(USRLIBDIR)/lint
+          MANPATH = /usr/X11R6/man
+    MANSOURCEPATH = $(MANPATH)/man
+           MANDIR = $(MANSOURCEPATH)1
+        LIBMANDIR = $(MANSOURCEPATH)3
+       FILEMANDIR = $(MANSOURCEPATH)5
+       MISCMANDIR = $(MANSOURCEPATH)$(MISCMANSUFFIX)
+     DRIVERMANDIR = $(MANSOURCEPATH)$(DRIVERMANSUFFIX)
+     LOGDIRECTORY = /var/log
+
+        VARRUNDIR = /var/run
+
+         VARDBDIR = /var/state
+
+               AR = ar clq
+  BOOTSTRAPCFLAGS =
+               CC = gcc
+               AS = as
+
+.SUFFIXES: .cc
+
+              CXX = c++
+          CXXFILT = c++filt
+           CXXLIB =
+    CXXDEBUGFLAGS = -O2 -fno-strength-reduce
+CXXDEPENDINCLUDES =
+ CXXEXTRA_DEFINES =
+CXXEXTRA_INCLUDES =
+   CXXSTD_DEFINES = -Dlinux -D__i386__ -D_POSIX_C_SOURCE=199309L -D_POSIX_SOURCE -D_XOPEN_SOURCE -D_BSD_SOURCE -D_SVID_SOURCE  -D_GNU_SOURCE  $(CXXPROJECT_DEFINES)
+       CXXOPTIONS =
+      CXXINCLUDES = $(INCLUDES) $(TOP_INCLUDES) $(CXXEXTRA_INCLUDES)
+       CXXDEFINES = $(CXXINCLUDES) $(CXXSTD_DEFINES) $(THREADS_CXXDEFINES) $(CXXEXTRA_DEFINES) $(DEFINES)
+         CXXFLAGS = $(CXXDEBUGFLAGS) $(CXXOPTIONS) $(THREADS_CXXFLAGS) $(CXXDEFINES)
+
+         COMPRESS = compress
+          GZIPCMD = gzip
+              CPP = /lib/cpp $(STD_CPP_DEFINES)
+           RAWCPP = /lib/cpp -undef $(STD_CPP_OPTIONS)
+    PREPROCESSCMD = gcc -E $(STD_CPP_DEFINES)
+          INSTALL = install
+     INSTALLFLAGS = -c
+               LD = ld
+              LEX = flex -l
+           LEXLIB = -lfl
+             YACC = bison -y
+           CCYACC = bison -y
+             LINT = lint
+      LINTLIBFLAG = -C
+         LINTOPTS = -axz
+               LN = ln -s
+             MAKE = make
+               MV = mv -f
+               CP = cp
+
+           RANLIB = ranlib
+  RANLIBINSTFLAGS =
+
+               RM = rm -f
+        MANSUFFIX = 1x
+     LIBMANSUFFIX = 3x
+    FILEMANSUFFIX = 5x
+    MISCMANSUFFIX = 7
+  DRIVERMANSUFFIX = 4
+          MANDEFS = -D__filemansuffix__=$(FILEMANSUFFIX) -D__miscmansuffix__=$(MISCMANSUFFIX) -D__drivermansuffix__=$(DRIVERMANSUFFIX)
+
+            TROFF = groff -Tps
+            NROFF = nroff
+         MSMACROS = -ms
+        MANMACROS = -man
+              TBL = tbl
+              EQN = eqn
+             NEQN = neqn
+              COL = col
+         COLFLAGS = -b
+            MODCC = gcc
+           MODCPP = /lib/cpp
+        MODCFLAGS = $(CDEBUGFLAGS) $(CCOPTIONS) $(THREAD_CFLAGS) $(ALLDEFINES)
+            MODAS = as
+       MODASFLAGS =
+            MODLD = ld
+       MODLDFLAGS =
+MODLDCOMBINEFLAGS = -r
+            MODAR = ar clq
+        MODRANLIB =
+
+            DVIPS = dvips
+            LATEX = latex
+
+     STD_INCLUDES =
+  STD_CPP_OPTIONS = -traditional
+  STD_CPP_DEFINES = -traditional -Dlinux -D__i386__ -D_POSIX_C_SOURCE=199309L -D_POSIX_SOURCE -D_XOPEN_SOURCE -D_BSD_SOURCE -D_SVID_SOURCE  -D_GNU_SOURCE  $(PROJECT_DEFINES)
+      STD_DEFINES = -Dlinux -D__i386__ -D_POSIX_C_SOURCE=199309L -D_POSIX_SOURCE -D_XOPEN_SOURCE -D_BSD_SOURCE -D_SVID_SOURCE  -D_GNU_SOURCE  $(PROJECT_DEFINES)
+ EXTRA_LOAD_FLAGS = -Wl,-rpath-link,$(BUILDLIBDIR)
+  EXTRA_LDOPTIONS =
+  EXTRA_LIBRARIES =
+             TAGS = ctags
+
+   PARALLELMFLAGS =
+
+    SHAREDCODEDEF =
+         SHLIBDEF =
+
+     SHLIBLDFLAGS = -shared
+
+         PICFLAGS = -fPIC
+
+      CXXPICFLAGS = -fPIC
+
+    PROTO_DEFINES = -DFUNCPROTO=15 -DNARROWPROTO
+
+     INSTPGMFLAGS = -s
+
+     INSTBINFLAGS = -m 0755
+     INSTUIDFLAGS = -m 4711
+     INSTLIBFLAGS = -m 0644
+     INSTINCFLAGS = -m 0444
+     INSTMANFLAGS = -m 0444
+     INSTDATFLAGS = -m 0444
+    INSTKMEMFLAGS = -m 4711
+
+      PROJECTROOT = /usr/X11R6
+
+      CDEBUGFLAGS = -O2 -fno-strength-reduce
+        CCOPTIONS = -ansi -pedantic -Wall -Wpointer-arith -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wredundant-decls -Wnested-externs
+
+      ALLINCLUDES = $(INCLUDES) $(EXTRA_INCLUDES) $(TOP_INCLUDES) $(STD_INCLUDES)
+       ALLDEFINES = $(ALLINCLUDES) $(STD_DEFINES) $(EXTRA_DEFINES) $(PROTO_DEFINES) $(THREADS_DEFINES) $(MODULE_DEFINES) $(DEFINES)
+           CFLAGS = $(CDEBUGFLAGS) $(CCOPTIONS) $(THREADS_CFLAGS) $(MODULE_CFLAGS) $(ALLDEFINES)
+        LINTFLAGS = $(LINTOPTS) -DLINT $(ALLDEFINES) $(DEPEND_DEFINES)
+         LDPRELIB = -L$(BUILDLIBDIR)
+        LDPOSTLIB =
+        LDOPTIONS = $(CDEBUGFLAGS) $(CCOPTIONS)  $(EXTRA_LDOPTIONS) $(THREADS_LDFLAGS) $(LOCAL_LDFLAGS) $(LDPRELIBS)
+     CXXLDOPTIONS = $(CXXDEBUGFLAGS) $(CXXOPTIONS) $(EXTRA_LDOPTIONS) $(THREADS_CXXLDFLAGS) $(LOCAL_LDFLAGS) $(LDPRELIBS)
+
+           LDLIBS = $(LDPOSTLIBS) $(THREADS_LIBS) $(SYS_LIBRARIES) $(EXTRA_LIBRARIES)
+
+           CCLINK = $(CC)
+
+          CXXLINK = $(CXX)
+
+     LDSTRIPFLAGS = -x
+   LDCOMBINEFLAGS = -r
+      DEPENDFLAGS =
+
+# Not sure this belongs here
+         TKLIBDIR =
+         TKINCDIR =
+        TKLIBNAME =
+        TKLIBRARY = -L$(TKLIBDIR) -l$(TKLIBNAME)
+        TCLLIBDIR =
+        TCLINCDIR =
+       TCLLIBNAME =
+       TCLLIBRARY = -L$(TCLLIBDIR) -l$(TCLLIBNAME)
+
+        MACROFILE = linux.cf
+           RM_CMD = $(RM)
+
+    IMAKE_DEFINES =
+
+         IRULESRC = $(CONFIGSRC)/cf
+        IMAKE_CMD = $(IMAKE) -I$(IRULESRC) $(IMAKE_DEFINES)
+
+     ICONFIGFILES = $(IRULESRC)/Imake.tmpl $(IRULESRC)/X11.tmpl 			$(IRULESRC)/site.def $(IRULESRC)/$(MACROFILE) 			$(IRULESRC)/xfree86.cf $(IRULESRC)/xf86.rules $(IRULESRC)/xf86site.def $(IRULESRC)/host.def $(EXTRA_ICONFIGFILES)
+
+# $TOG: X11.rules /main/4 1997/04/30 15:23:24 kaleb $
+
+# $XFree86: xc/config/cf/X11.rules,v 1.4 1997/09/09 11:56:20 hohndel Exp $
+
+# ----------------------------------------------------------------------
+# X Window System Build Parameters and Rules
+# $TOG: X11.tmpl /main/308 1998/06/16 15:14:24 msr $
+#
+#
+#
+#
+# $XFree86: xc/config/cf/X11.tmpl,v 1.59 2000/02/18 16:23:07 dawes Exp $
+
+STICKY_DEFINES = -DHAS_STICKY_DIR_BIT
+
+FCHOWN_DEFINES = -DHAS_FCHOWN
+
+# -----------------------------------------------------------------------
+# X Window System make variables; these need to be coordinated with rules
+
+             XTOP = $(TOP)
+           BINDIR = /usr/X11R6/bin
+     BUILDINCROOT = $(TOP)/exports
+      BUILDINCDIR = $(BUILDINCROOT)/include
+      BUILDINCTOP = ../..
+      BUILDLIBDIR = $(TOP)/exports/lib
+      BUILDLIBTOP = ../..
+      BUILDBINDIR = $(TOP)/exports/bin
+      BUILDBINTOP = ../..
+   BUILDMODULEDIR = $(BUILDLIBDIR)/modules
+   BUILDMODULETOP = $(BUILDLIBTOP)/..
+    XBUILDINCROOT = $(XTOP)/exports
+     XBUILDINCDIR = $(XBUILDINCROOT)/include/X11
+     XBUILDINCTOP = ../../..
+     XBUILDBINDIR = $(XBUILDINCROOT)/bin
+           INCDIR = $(INCROOT)
+           ADMDIR = /usr/adm
+           LIBDIR = $(USRLIBDIR)/X11
+       LIBEXECDIR = /usr/X11R6/libexec
+        MODULEDIR = $(USRLIBDIR)/modules
+   TOP_X_INCLUDES = -I$(TOP)/exports/include
+
+           VARDIR = /var/X11
+
+        ETCX11DIR = /etc/X11
+
+          CONFDIR = $(ETCX11DIR)
+
+           DOCDIR = $(LIBDIR)/doc
+       DOCHTMLDIR = $(DOCDIR)/html
+         DOCPSDIR = $(DOCDIR)/PostScript
+          FONTDIR = $(LIBDIR)/fonts
+     ENCODINGSDIR = $(FONTDIR)/encodings
+         XINITDIR = $(LIBDIR)/xinit
+           XDMDIR = $(LIBDIR)/xdm
+        XDMVARDIR = $(VARLIBDIR)/xdm
+           TWMDIR = $(LIBDIR)/twm
+           XSMDIR = $(LIBDIR)/xsm
+           NLSDIR = $(LIBDIR)/nls
+       XLOCALEDIR = $(LIBDIR)/locale
+        PEXAPIDIR = $(LIBDIR)/PEX
+      LBXPROXYDIR = $(LIBDIR)/lbxproxy
+  PROXYMANAGERDIR = $(LIBDIR)/proxymngr
+        XPRINTDIR = $(LIBDIR)/xserver
+      XAPPLOADDIR = $(LIBDIR)/app-defaults
+       FONTCFLAGS = -t
+
+     INSTAPPFLAGS = $(INSTDATFLAGS)
+
+              RGB = $(CLIENTENVSETUP) $(XBUILDBINDIR)/rgb
+            FONTC = $(CLIENTENVSETUP) $(XBUILDBINDIR)/bdftopcf
+        MKFONTDIR = $(CLIENTENVSETUP) $(XBUILDBINDIR)/mkfontdir
+
+       DOCUTILSRC = $(XTOP)/doc/util
+        CLIENTSRC = $(TOP)/clients
+          DEMOSRC = $(TOP)/demos
+       XDOCMACROS = $(DOCUTILSRC)/macros.t
+       XIDXMACROS = $(DOCUTILSRC)/indexmacros.t
+       PROGRAMSRC = $(TOP)/programs
+           LIBSRC = $(XTOP)/lib
+          FONTSRC = $(XTOP)/fonts
+     ENCODINGSSRC = $(FONTSRC)/encodings
+       INCLUDESRC = $(BUILDINCROOT)/include
+      XINCLUDESRC = $(INCLUDESRC)/X11
+        SERVERSRC = $(XTOP)/programs/Xserver
+       CONTRIBSRC = $(XTOP)/../contrib
+   UNSUPPORTEDSRC = $(XTOP)/unsupported
+           DOCSRC = $(XTOP)/doc
+           RGBSRC = $(XTOP)/programs/rgb
+      BDFTOPCFSRC = $(PROGRAMSRC)/bdftopcf
+     MKFONTDIRSRC = $(PROGRAMSRC)/mkfontdir
+    FONTSERVERSRC = $(PROGRAMSRC)/xfs
+       FONTINCSRC = $(XTOP)/include/fonts
+        EXTINCSRC = $(XTOP)/include/extensions
+      FTSOURCEDIR = $(TOP)/extras/FreeType
+     XTTSOURCEDIR = $(TOP)/extras/X-TrueType
+     TRANSCOMMSRC = $(LIBSRC)/xtrans
+   TRANS_INCLUDES = -I$(TRANSCOMMSRC)
+ CONNECTION_FLAGS = -DUNIXCONN -DTCPCONN $(STICKY_DEFINES) $(FCHOWN_DEFINES)
+
+       XENVLIBDIR = $(TOP)/exports/lib
+   CLIENTENVSETUP = LD_LIBRARY_PATH=$(XENVLIBDIR)
+
+# $TOG: lnxLib.tmpl /main/7 1998/03/20 14:26:41 mgreess $
+# $XFree86: xc/config/cf/lnxLib.tmpl,v 3.12 1999/06/27 14:07:17 dawes Exp $
+
+          XLIBSRC = $(LIBSRC)/X11
+
+SOXLIBREV = 6.1
+DEPXONLYLIB =
+XONLYLIB =  -lX11
+
+LINTXONLY = $(XLIBSRC)/llib-X11.ln
+
+         XLIBONLY = $(XONLYLIB)
+
+      XEXTLIBSRC = $(LIBSRC)/Xext
+
+SOXEXTREV = 6.4
+DEPEXTENSIONLIB =
+EXTENSIONLIB =  -lXext
+
+LINTEXTENSION = $(XEXTLIBSRC)/llib-Xext.ln
+
+LINTEXTENSIONLIB = $(LINTEXTENSION)
+          DEPXLIB = $(DEPEXTENSIONLIB) $(DEPXONLYLIB)
+             XLIB = $(EXTENSIONLIB) $(XONLYLIB)
+         LINTXLIB = $(LINTXONLYLIB)
+
+    XSSLIBSRC = $(LIBSRC)/Xss
+
+DEPXSSLIB = $(TOP)/exports/lib/libXss.a
+XSSLIB =  -lXss
+
+LINTXSS = $(XSSLIBSRC)/llib-Xss.ln
+
+    XXF86MISCLIBSRC = $(LIBSRC)/Xxf86misc
+
+DEPXXF86MISCLIB = $(TOP)/exports/lib/libXxf86misc.a
+XXF86MISCLIB =  -lXxf86misc
+
+LINTXXF86MISC = $(XXF86MISCLIBSRC)/llib-Xxf86misc.ln
+
+    XXF86VMLIBSRC = $(LIBSRC)/Xxf86vm
+
+DEPXXF86VMLIB = $(TOP)/exports/lib/libXxf86vm.a
+XXF86VMLIB =  -lXxf86vm
+
+LINTXXF86VM = $(XXF86VMLIBSRC)/llib-Xxf86vm.ln
+
+    XXF86DGALIBSRC = $(LIBSRC)/Xxf86dga
+
+DEPXXF86DGALIB = $(TOP)/exports/lib/libXxf86dga.a
+XXF86DGALIB =  -lXxf86dga
+
+LINTXXF86DGA = $(XXF86DGALIBSRC)/llib-Xxf86dga.ln
+
+    XXF86RUSHLIBSRC = $(LIBSRC)/Xxf86rush
+
+DEPXXF86RUSHLIB = $(TOP)/exports/lib/libXxf86rush.a
+XXF86RUSHLIB =  -lXxf86rush
+
+LINTXXF86RUSH = $(XXF86RUSHLIBSRC)/llib-Xxf86rush.ln
+
+    XVLIBSRC = $(LIBSRC)/Xv
+
+DEPXVLIB = $(TOP)/exports/lib/libXv.a
+XVLIB =  -lXv
+
+LINTXV = $(XVLIBSRC)/llib-Xv.ln
+
+    DPSLIBSRC = $(LIBSRC)/dps
+
+SODPSREV = 1.0
+DEPDPSLIB =
+DPSLIB =  -ldps
+
+LINTDPS = $(DPSLIBSRC)/llib-dps.ln
+
+    DPSTKLIBSRC = $(LIBSRC)/dpstk
+
+SODPSTKREV = 1.0
+DEPDPSTKLIB =
+DPSTKLIB =  -ldpstk
+
+LINTDPSTK = $(DPSTKLIBSRC)/llib-dpstk.ln
+
+    PSRESLIBSRC = $(LIBSRC)/psres
+
+SOPSRESREV = 1.0
+DEPPSRESLIB =
+PSRESLIB =  -lpsres
+
+LINTPSRES = $(PSRESLIBSRC)/llib-psres.ln
+
+    GLXLIBSRC = $(LIBSRC)/GL
+
+SOGLREV = 1.2
+DEPGLXLIB =
+GLXLIB =  -lGL
+
+LINTGLX = $(GLXLIBSRC)/llib-GL.ln
+
+    XFONTCACHELIBSRC = $(LIBSRC)/Xfontcache
+
+DEPXFONTCACHELIB = $(TOP)/exports/lib/libXfontcache.a
+XFONTCACHELIB =  -lXfontcache
+
+LINTXFONTCACHE = $(XFONTCACHELIBSRC)/llib-Xfontcache.ln
+
+         XAUTHSRC = $(LIBSRC)/Xau
+
+DEPXAUTHLIB = $(TOP)/exports/lib/libXau.a
+XAUTHLIB =  -lXau
+
+LINTXAUTH = $(XAUTHSRC)/llib-Xau.ln
+
+      XDMCPLIBSRC = $(LIBSRC)/Xdmcp
+
+DEPXDMCPLIB = $(TOP)/exports/lib/libXdmcp.a
+XDMCPLIB =  -lXdmcp
+
+LINTXDMCP = $(XDMCPLIBSRC)/llib-Xdmcp.ln
+
+           XMUSRC = $(LIBSRC)/Xmu
+
+SOXMUREV = 6.1
+DEPXMULIB =
+XMULIB =  -lXmu
+
+LINTXMU = $(XMUSRC)/llib-Xmu.ln
+
+       OLDXLIBSRC = $(LIBSRC)/oldX
+
+DEPOLDXLIB = $(TOP)/exports/lib/liboldX.a
+OLDXLIB =  -loldX
+
+LINTOLDX = $(OLDXLIBSRC)/llib-oldX.ln
+
+         XPLIBSRC = $(LIBSRC)/Xp
+
+SOXPREV = 6.2
+DEPXPLIB =
+XPLIB =  -lXp
+
+LINTXP = $(XPLIBSRC)/llib-Xp.ln
+
+       TOOLKITSRC = $(LIBSRC)/Xt
+
+SOXTREV = 6.0
+DEPXTOOLONLYLIB =
+XTOOLONLYLIB =  -lXt
+
+LINTXTOOLONLY = $(TOOLKITSRC)/llib-Xt.ln
+
+      DEPXTOOLLIB = $(DEPXTOOLONLYLIB) $(DEPSMLIB) $(DEPICELIB)
+         XTOOLLIB = $(XTOOLONLYLIB) $(SMLIB) $(ICELIB)
+     LINTXTOOLLIB = $(LINTXTOOLONLYLIB)
+
+       XALIBSRC = $(LIBSRC)/Xa
+
+SOXAREV = 1.0
+DEPXALIB =
+XALIB =  -lXa
+
+LINTXA = $(XALIBSRC)/llib-Xa.ln
+
+       AWIDGETSRC = $(LIBSRC)/Xaw
+
+SOXAWREV = 7.0
+DEPXAWLIB =
+XAWLIB =  -lXaw
+
+LINTXAW = $(AWIDGETSRC)/llib-Xaw.ln
+
+       AWIDGET6SRC = $(LIBSRC)/Xaw6
+
+SOXAW6REV = 6.1
+DEPXAW6LIB =
+XAW6LIB =  -lXaw
+
+LINTXAW6 = $(AWIDGET6SRC)/llib-Xaw.ln
+
+         XILIBSRC = $(LIBSRC)/Xi
+
+SOXINPUTREV = 6.0
+DEPXILIB =
+XILIB =  -lXi
+
+LINTXI = $(XILIBSRC)/llib-Xi.ln
+
+      XTESTLIBSRC = $(LIBSRC)/Xtst
+
+SOXTESTREV = 6.1
+DEPXTESTLIB =
+XTESTLIB =  -lXtst
+
+LINTXTEST = $(XTESTLIBSRC)/llib-Xtst.ln
+
+        PEXLIBSRC = $(LIBSRC)/PEX5
+
+SOPEXREV = 6.0
+DEPPEXLIB =
+PEXLIB =  -lPEX5
+
+LINTPEX = $(PEXLIBSRC)/llib-PEX5.ln
+
+        XIELIBSRC = $(LIBSRC)/XIE
+
+SOXIEREV = 6.0
+DEPXIELIB =
+XIELIB =  -lXIE
+
+LINTXIE = $(XIELIBSRC)/llib-XIE.ln
+
+      PHIGSLIBSRC = $(LIBSRC)/PHIGS
+
+DEPPHIGSLIB = $(TOP)/exports/lib/libphigs.a
+PHIGSLIB =  -lphigs
+
+LINTPHIGS = $(PHIGSLIBSRC)/llib-phigs.ln
+
+DEPXBSDLIB = $(TOP)/exports/lib/libXbsd.a
+XBSDLIB =  -lXbsd
+
+LINTXBSD = $(LIBSRC)/Xbsd/llib-Xbsd.ln
+
+           ICESRC = $(LIBSRC)/ICE
+
+SOICEREV = 6.3
+DEPICELIB =
+ICELIB =  -lICE
+
+LINTICE = $(ICESRC)/llib-ICE.ln
+
+            SMSRC = $(LIBSRC)/SM
+
+SOSMREV = 6.0
+DEPSMLIB =
+SMLIB =  -lSM
+
+LINTSM = $(SMSRC)/llib-SM.ln
+
+           XKEYSRC = $(LIBSRC)/Xkey
+
+SOXKEYREV = 6.0
+DEPXKEYLIB =
+XKEYLIB =  -lXkey
+
+LINTXKEY = $(XKEYSRC)/llib-Xkey.ln
+
+         FSLIBSRC = $(LIBSRC)/FS
+
+DEPFSLIB = $(TOP)/exports/lib/libFS.a
+FSLIB =  -lFS
+
+LINTFS = $(FSLIBSRC)/llib-FS.ln
+
+         FONTLIBSRC = $(LIBSRC)/font
+
+SOFONTREV = 1.3
+DEPFONTLIB =
+FONTLIB = -L$(FREETYPELIBDIR) -L$(FONTLIBSRC)  -lXfont
+
+LINTXFONT = $(FONTLIBSRC)/llib-Xfont.ln
+#
+SOFONTREV = 1.3
+DEPXFONTLIB =
+XFONTLIB =  -lXfont
+
+LINTXFONT = $(FONTLIBSRC)/llib-Xfont.ln
+
+     FONTSTUBLIBSRC = $(FONTLIBSRC)/stubs
+
+DEPFONTSTUBLIB = $(TOP)/exports/lib/libfntstubs.a
+FONTSTUBLIB =  -lfntstubs
+
+LINTFONTSTUB = $(FONTSUBLIBSRC)/llib-fntstubs.ln
+         DEPFONTLIB = $(DEPXFONTLIB) $(DEPFONTSTUBLIB)
+            FONTLIB = $(XFONTLIB) $(FONTSTUBLIB)
+
+          XPMLIBSRC = $(LIBSRC)/Xpm
+
+SOXPMREV = 4.11
+DEPXPMLIB =
+XPMLIB =  -lXpm
+
+LINTXPM = $(XPMLIBSRC)/llib-Xpm.ln
+
+    XKBFILELIBSRC = $(LIBSRC)/xkbfile
+
+DEPXKBFILELIB = $(TOP)/exports/lib/libxkbfile.a
+XKBFILELIB =  -lxkbfile
+
+LINTXKBFILE = $(XKBFILELIBSRC)/llib-xkbfile.ln
+
+     XKBCOMPSRC = $(PROGRAMSRC)/xkbcomp
+     XKBCOMPCMD = $(CLIENTENVSETUP) $(XBUILDBINDIR)/xkbcomp
+
+    XKBUILIBSRC = $(LIBSRC)/xkbui
+
+DEPXKBUILIB = $(TOP)/exports/lib/libxkbui.a
+XKBUILIB =  -lxkbui
+
+LINTXKBUI = $(XKBUILIBSRC)/llib-xkbui.ln
+
+          DEPLIBS = $(DEPXAWLIB) $(DEPXMULIB) $(DEPXTOOLLIB) $(DEPXLIB)
+
+         DEPLIBS1 = $(DEPLIBS)
+         DEPLIBS2 = $(DEPLIBS)
+         DEPLIBS3 = $(DEPLIBS)
+         DEPLIBS4 = $(DEPLIBS)
+         DEPLIBS5 = $(DEPLIBS)
+         DEPLIBS6 = $(DEPLIBS)
+         DEPLIBS7 = $(DEPLIBS)
+         DEPLIBS8 = $(DEPLIBS)
+         DEPLIBS9 = $(DEPLIBS)
+         DEPLIBS10 = $(DEPLIBS)
+
+XMULIBONLY = -lXmu
+XMULIB = $(XMULIBONLY) $(XTOOLLIB) $(XLIB)
+
+        CONFIGDIR = $(LIBDIR)/config
+
+    USRLIBDIRPATH = $(USRLIBDIR)
+        LDPRELIBS = -L$(BUILDLIBDIR)
+       LDPOSTLIBS =
+     TOP_INCLUDES = -I$(TOP) $(TOP_X_INCLUDES)
+  PROJECT_DEFINES =
+
+CXXPROJECT_DEFINES =
+
+# ----------------------------------------------------------------------
+# start of Imakefile
+
+# $TOG: Imakefile /main/10 1997/04/30 15:39:17 kaleb $
+#
+#
+#
+#
+# $XFree86: xc/config/imake/Imakefile,v 3.9 1999/08/14 10:49:01 dawes Exp $
+
+# Some compilers generate fatal errors if an -L directory does
+# not exist.  Since BUILDLIBDIR may not exist yet suppress its use.
+         LDPRELIB =
+        LDPRELIBS =
+
+DEPLIBS =
+
+INCLUDES = -I$(TOP)/include $(TOP_X_INCLUDES)/X11
+CPP_PROGRAM = /lib/cpp
+CPP_DEFINES = -DCPP_PROGRAM="\"$(CPP_PROGRAM)\""
+DEFINES = $(SIGNAL_DEFINES) $(BOOTSTRAPCFLAGS) $(CPP_DEFINES)
+
+ OBJS = imake.o
+ SRCS = imake.c
+
+ PROGRAM = imake
+
+all:: imake
+
+imake: $(OBJS) $(DEPLIBS)
+	$(RM) $@
+	$(CCLINK) -o $@ $(LDOPTIONS) $(OBJS) $(LOCAL_LIBRARIES) $(LDLIBS)  $(EXTRA_LOAD_FLAGS)
+
+install:: imake
+	@if [ -d $(DESTDIR)$(BINDIR) ]; then set +x; \
+	else (set -x; $(MKDIRHIER) $(DESTDIR)$(BINDIR)); fi
+	$(INSTALL) $(INSTALLFLAGS) $(INSTPGMFLAGS)  imake $(DESTDIR)$(BINDIR)/imake
+
+install.man:: imake.man
+	@if [ -d $(DESTDIR)$(MANDIR) ]; then set +x; \
+	else (set -x; $(MKDIRHIER) $(DESTDIR)$(MANDIR)); fi
+	$(INSTALL) $(INSTALLFLAGS) $(INSTMANFLAGS) imake.man $(DESTDIR)$(MANDIR)/imake.$(MANSUFFIX)
+
+depend:: $(DEPEND)
+
+$(DEPEND):
+	@echo "checking $@ over in $(DEPENDSRC) first..."; \
+	cd $(DEPENDSRC) && $(MAKE); \
+	echo "okay, continuing in $(CURRENT_DIR)"
+
+depend::
+	$(DEPEND) $(DEPENDFLAGS) -- $(ALLDEFINES) $(DEPEND_DEFINES) -- $(SRCS)
+
+lint:
+	$(LINT) $(LINTFLAGS) $(SRCS) $(LINTLIBS)
+lint1:
+	$(LINT) $(LINTFLAGS) $(FILE) $(LINTLIBS)
+
+clean::
+	$(RM) imake
+
+install.sdk:: imake
+	@if [ -d $(DESTDIR)$(DRIVERSDKDIR)/config/imake ]; then set +x; \
+	else (set -x; $(MKDIRHIER) $(DESTDIR)$(DRIVERSDKDIR)/config/imake); fi
+	$(INSTALL) -c $(INSTBINFLAGS) imake $(DESTDIR)$(DRIVERSDKDIR)/config/imake/imake
+
+bootstrapdepend:
+
+clean::
+	$(RM) ccimake
+	$(RM) -r bootstrap
+
+# ----------------------------------------------------------------------
+# common rules for all Makefiles - do not edit
+
+.c.i:
+	$(RM) $@
+	 $(CC) -E $(CFLAGS) $(_NOOP_) $*.c > $@
+
+.SUFFIXES: .s
+
+.c.s:
+	$(RM) $@
+	 $(CC) -S $(CFLAGS) $(_NOOP_) $*.c
+
+emptyrule::
+
+clean::
+	$(RM) *.CKP *.ln *.BAK *.bak *.o core errs ,* *~ *.a .emacs_* tags TAGS make.log MakeOut  "#"*
+
+Makefile::  imake
+
+Makefile::
+	-@if [ -f Makefile ]; then set -x; \
+	$(RM) Makefile.bak; $(MV) Makefile Makefile.bak; \
+	else exit 0; fi
+	$(IMAKE_CMD) -DTOPDIR=$(TOP) -DCURDIR=$(CURRENT_DIR)
+
+tags::
+	$(TAGS) -w *.[ch]
+	$(TAGS) -xw *.[ch] > TAGS
+
+man_keywords::
+
+# ----------------------------------------------------------------------
+# empty rules for directories that do not have SUBDIRS - do not edit
+
+install::
+	@echo "install in $(CURRENT_DIR) done"
+
+install.man::
+	@echo "install.man in $(CURRENT_DIR) done"
+
+install.sdk::
+	@echo "install.sdk in $(CURRENT_DIR) done"
+
+Makefiles::
+
+includes::
+
+depend::
+
+# ----------------------------------------------------------------------
+# dependencies generated by makedepend
+
diff -urN XFree86-4.0.orig/xc/config/imake/imake.c XFree86-4.0/xc/config/imake/imake.c
--- XFree86-4.0.orig/xc/config/imake/imake.c	Wed Dec 29 01:54:48 1999
+++ XFree86-4.0/xc/config/imake/imake.c	Fri Mar 17 02:19:49 2000
@@ -934,6 +934,7 @@
   static char* yast = "/sbin/YaST";
   static char* redhat = "/etc/redhat-release";
   static char* debian = "/etc/debian_version";
+  static char* pld = "/etc/pld-release";
 
   fprintf (inFile, "%s\n", "#define LinuxUnknown    0");
   fprintf (inFile, "%s\n", "#define LinuxSuSE       1");
@@ -948,6 +949,7 @@
   fprintf (inFile, "%s\n", "#define LinuxTurbo      10");
   fprintf (inFile, "%s\n", "#define LinuxWare       11");
   fprintf (inFile, "%s\n", "#define LinuxYggdrasil  12");
+  fprintf (inFile, "%s\n", "#define LinuxPLD        13");
 
   if (lstat (yast, &sb) == 0) {
     fprintf (inFile, "%s\n", "#define DefaultLinuxDistribution LinuxSuSE");
@@ -964,6 +966,11 @@
     fprintf (inFile, "%s\n", "#define DefaultLinuxDistName Debian");
     /* You could also try to get the version of the Debian distrib by looking
      * at the content of /etc/debian_version */
+    return;
+  }
+  if (lstat (pld, &sb) == 0) {
+    fprintf (inFile, "%s\n", "#define DefaultLinuxDistribution LinuxPLD");
+    fprintf (inFile, "%s\n", "#define DefaultLinuxDistName PLD");
     return;
   }
   /* what's the definitive way to tell what any particular distribution is? */
diff -urN XFree86-4.0.orig/xc/config/makedepend/Makefile.proto XFree86-4.0/xc/config/makedepend/Makefile.proto
--- XFree86-4.0.orig/xc/config/makedepend/Makefile.proto	Thu Jan  1 01:00:00 1970
+++ XFree86-4.0/xc/config/makedepend/Makefile.proto	Fri Mar 17 02:19:49 2000
@@ -0,0 +1,890 @@
+# Makefile generated by imake - do not edit!
+# $TOG: imake.c /main/104 1998/03/24 12:45:15 kaleb $
+
+# ----------------------------------------------------------------------
+# Makefile generated from "Imake.tmpl" and <./config/makedepend/Imakefile>
+# $TOG: Imake.tmpl /main/249 1997/10/13 15:28:56 kaleb $
+#
+#
+#
+#
+# $XFree86: xc/config/cf/Imake.tmpl,v 3.65 2000/01/31 19:33:18 dawes Exp $
+# ----------------------------------------------------------------------
+
+all::
+
+.SUFFIXES: .i
+
+# $TOG: Imake.cf /main/30 1998/04/28 13:55:25 barstow $
+
+# $XFree86: xc/config/cf/Imake.cf,v 3.53 2000/02/12 20:45:01 dawes Exp $
+
+# -----------------------------------------------------------------------
+# site-specific configuration parameters that need to come before
+# the platform-specific parameters - edit site.def to change
+
+# site:  $TOG: site.sample /main/r64_final/1 1998/02/05 16:28:49 kaleb $
+
+# site:  $XFree86: xc/config/cf/site.def,v 3.22 1998/10/02 06:15:06 dawes Exp $
+
+# $XFree86: xc/config/cf/xf86site.def,v 3.153 1999/12/03 19:17:17 eich Exp $
+
+# ----------------------------------------------------------------------
+# platform-specific configuration parameters - edit linux.cf to change
+
+# platform:  $TOG: linux.cf /main/47 1998/04/17 11:32:51 mgreess
+# platform:  $XFree86: xc/config/cf/linux.cf,v 3.115 2000/02/21 18:05:31 dawes Exp $
+
+# operating system:  Linux 2.2.14 i586 [ELF] (2.2.14)
+# libc:	(6.1.2)
+# binutils:	(29)
+
+# $TOG: lnxLib.rules /main/14 1998/03/16 11:15:30 mgreess $
+# $XFree86: xc/config/cf/lnxLib.rules,v 3.37 1999/06/06 08:47:44 dawes Exp $
+
+# $XFree86: xc/config/cf/xfree86.cf,v 3.269 2000/02/17 13:50:02 dawes Exp $
+
+# $TOG: xfree86.cf /main/38 1997/10/22 13:01:59 kaleb $
+
+AFB_DEFS = -DUSE_AFB
+
+DRIVERSDKDIR = $(USRLIBDIR)/Server
+DRIVERSDKMODULEDIR = $(USRLIBDIR)/Server/modules
+DRIVERSDKINCLUDEDIR = $(USRLIBDIR)/Server/include
+
+       XF86SRC = $(SERVERSRC)/hw/xfree86
+  XF86ACCELSRC = $(XF86SRC)/accel
+    XF86COMSRC = $(XF86SRC)/common
+ XF86PARSERSRC = $(XF86SRC)/parser
+     XF86OSSRC = $(XF86SRC)/os-support
+ XF86DRIVERSRC = $(XF86SRC)/drivers
+     DRIVERSRC = $(XF86DRIVERSRC)
+   S3DRIVERSRC = $(XF86SRC)/accel/s3/drivers
+  S3VDRIVERSRC = $(XF86SRC)/accel/s3_virge/drivers
+
+       XF68SRC = $(SERVERSRC)/hw/xfree68
+    XF68COMSRC = $(XF68SRC)/common
+ XF68PARSERSRC = $(XF68SRC)/parser
+     XF68OSSRC = $(XF68SRC)/os-support
+
+           XF98SRC = $(SERVERSRC)/hw/xfree98
+      XF98ACCELSRC = $(XF98SRC)/accel
+        XF98COMSRC = $(XF98SRC)/common
+         XF98OSSRC = $(XF98SRC)/os-support
+     XF98DRIVERSRC = $(XF98SRC)/drivers
+  XF98VGADRIVERSRC = $(XF98SRC)/vga256/drivers
+XF98VGA16DRIVERSRC = $(XF98SRC)/vga16/drivers
+ XF98VGA2DRIVERSRC = $(XF98SRC)/vga2/drivers
+ XF98MONODRIVERSRC = $(XF98SRC)/mono/drivers
+XF98NECS3DRIVERSRC = $(XF98SRC)/accel/s3nec/drivers
+XF98PWSKBDRIVERSRC = $(XF98SRC)/accel/s3pwskb/drivers
+ XF98PWLBDRIVERSRC = $(XF98SRC)/accel/s3pwlb/drivers
+XF98GA968DRIVERSRC = $(XF98SRC)/accel/s3ga968/drivers
+
+        XFREE86DOCDIR = $(DOCDIR)
+      XFREE86PSDOCDIR = $(DOCPSDIR)
+    XFREE86HTMLDOCDIR = $(DOCHTMLDIR)
+XFREE86JAPANESEDOCDIR = $(DOCDIR)/Japanese
+
+# $XConsortium: xf86.rules /main/9 1996/10/31 14:54:26 kaleb $
+# $XFree86: xc/config/cf/xf86.rules,v 3.30 1999/10/13 16:49:02 dawes Exp $
+
+# ----------------------------------------------------------------------
+# site-specific configuration parameters that go after
+# the platform-specific parameters - edit site.def to change
+
+# site:  $TOG: site.sample /main/r64_final/1 1998/02/05 16:28:49 kaleb $
+
+# site:  $XFree86: xc/config/cf/site.def,v 3.22 1998/10/02 06:15:06 dawes Exp $
+
+# ---------------------------------------------------------------------
+# Imake rules for building libraries, programs, scripts, and data files
+# rules:  $TOG: Imake.rules /main/227 1998/02/02 12:07:57 kaleb $
+#
+#
+#
+#
+# rules:  $XFree86: xc/config/cf/Imake.rules,v 3.64 2000/02/13 06:15:16 dawes Exp $
+
+ _NULLCMD_ = @ echo -n
+
+GLIDEINCDIR =
+
+TKLIBNAME =
+
+TKLIBDIR =
+
+TCLLIBNAME =
+
+TCLIBDIR =
+
+          PATHSEP = /
+            SHELL = /bin/sh
+
+              TOP = ../..
+      CURRENT_DIR = ./config/makedepend
+
+            IMAKE = $(IMAKESRC)/imake
+           DEPEND = $(DEPENDSRC)/makedepend
+        MKDIRHIER = mkdir -p
+          REVPATH = $(CONFIGSRC)/util/revpath
+    EXPORTLISTGEN =
+        CONFIGSRC = $(TOP)/config
+         IMAKESRC = $(CONFIGSRC)/imake
+        DEPENDSRC = $(CONFIGSRC)/makedepend
+
+          INCROOT = /usr/X11R6/include
+        USRLIBDIR = /usr/X11R6/lib
+        VARLIBDIR = /var/lib
+         SHLIBDIR = /usr/X11R6/lib
+       LINTLIBDIR = $(USRLIBDIR)/lint
+          MANPATH = /usr/X11R6/man
+    MANSOURCEPATH = $(MANPATH)/man
+           MANDIR = $(MANSOURCEPATH)1
+        LIBMANDIR = $(MANSOURCEPATH)3
+       FILEMANDIR = $(MANSOURCEPATH)5
+       MISCMANDIR = $(MANSOURCEPATH)$(MISCMANSUFFIX)
+     DRIVERMANDIR = $(MANSOURCEPATH)$(DRIVERMANSUFFIX)
+     LOGDIRECTORY = /var/log
+
+        VARRUNDIR = /var/run
+
+         VARDBDIR = /var/state
+
+               AR = ar clq
+  BOOTSTRAPCFLAGS =
+               CC = gcc
+               AS = as
+
+.SUFFIXES: .cc
+
+              CXX = c++
+          CXXFILT = c++filt
+           CXXLIB =
+    CXXDEBUGFLAGS = -O2 -fno-strength-reduce
+CXXDEPENDINCLUDES =
+ CXXEXTRA_DEFINES =
+CXXEXTRA_INCLUDES =
+   CXXSTD_DEFINES = -Dlinux -D__i386__ -D_POSIX_C_SOURCE=199309L -D_POSIX_SOURCE -D_XOPEN_SOURCE -D_BSD_SOURCE -D_SVID_SOURCE  -D_GNU_SOURCE  $(CXXPROJECT_DEFINES)
+       CXXOPTIONS =
+      CXXINCLUDES = $(INCLUDES) $(TOP_INCLUDES) $(CXXEXTRA_INCLUDES)
+       CXXDEFINES = $(CXXINCLUDES) $(CXXSTD_DEFINES) $(THREADS_CXXDEFINES) $(CXXEXTRA_DEFINES) $(DEFINES)
+         CXXFLAGS = $(CXXDEBUGFLAGS) $(CXXOPTIONS) $(THREADS_CXXFLAGS) $(CXXDEFINES)
+
+         COMPRESS = compress
+          GZIPCMD = gzip
+              CPP = /lib/cpp $(STD_CPP_DEFINES)
+           RAWCPP = /lib/cpp -undef $(STD_CPP_OPTIONS)
+    PREPROCESSCMD = gcc -E $(STD_CPP_DEFINES)
+          INSTALL = install
+     INSTALLFLAGS = -c
+               LD = ld
+              LEX = flex -l
+           LEXLIB = -lfl
+             YACC = bison -y
+           CCYACC = bison -y
+             LINT = lint
+      LINTLIBFLAG = -C
+         LINTOPTS = -axz
+               LN = ln -s
+             MAKE = make
+               MV = mv -f
+               CP = cp
+
+           RANLIB = ranlib
+  RANLIBINSTFLAGS =
+
+               RM = rm -f
+        MANSUFFIX = 1x
+     LIBMANSUFFIX = 3x
+    FILEMANSUFFIX = 5x
+    MISCMANSUFFIX = 7
+  DRIVERMANSUFFIX = 4
+          MANDEFS = -D__filemansuffix__=$(FILEMANSUFFIX) -D__miscmansuffix__=$(MISCMANSUFFIX) -D__drivermansuffix__=$(DRIVERMANSUFFIX)
+
+            TROFF = groff -Tps
+            NROFF = nroff
+         MSMACROS = -ms
+        MANMACROS = -man
+              TBL = tbl
+              EQN = eqn
+             NEQN = neqn
+              COL = col
+         COLFLAGS = -b
+            MODCC = gcc
+           MODCPP = /lib/cpp
+        MODCFLAGS = $(CDEBUGFLAGS) $(CCOPTIONS) $(THREAD_CFLAGS) $(ALLDEFINES)
+            MODAS = as
+       MODASFLAGS =
+            MODLD = ld
+       MODLDFLAGS =
+MODLDCOMBINEFLAGS = -r
+            MODAR = ar clq
+        MODRANLIB =
+
+            DVIPS = dvips
+            LATEX = latex
+
+     STD_INCLUDES =
+  STD_CPP_OPTIONS = -traditional
+  STD_CPP_DEFINES = -traditional -Dlinux -D__i386__ -D_POSIX_C_SOURCE=199309L -D_POSIX_SOURCE -D_XOPEN_SOURCE -D_BSD_SOURCE -D_SVID_SOURCE  -D_GNU_SOURCE  $(PROJECT_DEFINES)
+      STD_DEFINES = -Dlinux -D__i386__ -D_POSIX_C_SOURCE=199309L -D_POSIX_SOURCE -D_XOPEN_SOURCE -D_BSD_SOURCE -D_SVID_SOURCE  -D_GNU_SOURCE  $(PROJECT_DEFINES)
+ EXTRA_LOAD_FLAGS = -Wl,-rpath-link,$(BUILDLIBDIR)
+  EXTRA_LDOPTIONS =
+  EXTRA_LIBRARIES =
+             TAGS = ctags
+
+   PARALLELMFLAGS =
+
+    SHAREDCODEDEF =
+         SHLIBDEF =
+
+     SHLIBLDFLAGS = -shared
+
+         PICFLAGS = -fPIC
+
+      CXXPICFLAGS = -fPIC
+
+    PROTO_DEFINES = -DFUNCPROTO=15 -DNARROWPROTO
+
+     INSTPGMFLAGS = -s
+
+     INSTBINFLAGS = -m 0755
+     INSTUIDFLAGS = -m 4711
+     INSTLIBFLAGS = -m 0644
+     INSTINCFLAGS = -m 0444
+     INSTMANFLAGS = -m 0444
+     INSTDATFLAGS = -m 0444
+    INSTKMEMFLAGS = -m 4711
+
+      PROJECTROOT = /usr/X11R6
+
+      CDEBUGFLAGS = -O2 -fno-strength-reduce
+        CCOPTIONS = -ansi -pedantic -Wall -Wpointer-arith -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wredundant-decls -Wnested-externs
+
+      ALLINCLUDES = $(INCLUDES) $(EXTRA_INCLUDES) $(TOP_INCLUDES) $(STD_INCLUDES)
+       ALLDEFINES = $(ALLINCLUDES) $(STD_DEFINES) $(EXTRA_DEFINES) $(PROTO_DEFINES) $(THREADS_DEFINES) $(MODULE_DEFINES) $(DEFINES)
+           CFLAGS = $(CDEBUGFLAGS) $(CCOPTIONS) $(THREADS_CFLAGS) $(MODULE_CFLAGS) $(ALLDEFINES)
+        LINTFLAGS = $(LINTOPTS) -DLINT $(ALLDEFINES) $(DEPEND_DEFINES)
+         LDPRELIB = -L$(BUILDLIBDIR)
+        LDPOSTLIB =
+        LDOPTIONS = $(CDEBUGFLAGS) $(CCOPTIONS)  $(EXTRA_LDOPTIONS) $(THREADS_LDFLAGS) $(LOCAL_LDFLAGS) $(LDPRELIBS)
+     CXXLDOPTIONS = $(CXXDEBUGFLAGS) $(CXXOPTIONS) $(EXTRA_LDOPTIONS) $(THREADS_CXXLDFLAGS) $(LOCAL_LDFLAGS) $(LDPRELIBS)
+
+           LDLIBS = $(LDPOSTLIBS) $(THREADS_LIBS) $(SYS_LIBRARIES) $(EXTRA_LIBRARIES)
+
+           CCLINK = $(CC)
+
+          CXXLINK = $(CXX)
+
+     LDSTRIPFLAGS = -x
+   LDCOMBINEFLAGS = -r
+      DEPENDFLAGS =
+
+# Not sure this belongs here
+         TKLIBDIR =
+         TKINCDIR =
+        TKLIBNAME =
+        TKLIBRARY = -L$(TKLIBDIR) -l$(TKLIBNAME)
+        TCLLIBDIR =
+        TCLINCDIR =
+       TCLLIBNAME =
+       TCLLIBRARY = -L$(TCLLIBDIR) -l$(TCLLIBNAME)
+
+        MACROFILE = linux.cf
+           RM_CMD = $(RM)
+
+    IMAKE_DEFINES =
+
+         IRULESRC = $(CONFIGSRC)/cf
+        IMAKE_CMD = $(IMAKE) -I$(IRULESRC) $(IMAKE_DEFINES)
+
+     ICONFIGFILES = $(IRULESRC)/Imake.tmpl $(IRULESRC)/X11.tmpl 			$(IRULESRC)/site.def $(IRULESRC)/$(MACROFILE) 			$(IRULESRC)/xfree86.cf $(IRULESRC)/xf86.rules $(IRULESRC)/xf86site.def $(IRULESRC)/host.def $(EXTRA_ICONFIGFILES)
+
+# $TOG: X11.rules /main/4 1997/04/30 15:23:24 kaleb $
+
+# $XFree86: xc/config/cf/X11.rules,v 1.4 1997/09/09 11:56:20 hohndel Exp $
+
+# ----------------------------------------------------------------------
+# X Window System Build Parameters and Rules
+# $TOG: X11.tmpl /main/308 1998/06/16 15:14:24 msr $
+#
+#
+#
+#
+# $XFree86: xc/config/cf/X11.tmpl,v 1.59 2000/02/18 16:23:07 dawes Exp $
+
+STICKY_DEFINES = -DHAS_STICKY_DIR_BIT
+
+FCHOWN_DEFINES = -DHAS_FCHOWN
+
+# -----------------------------------------------------------------------
+# X Window System make variables; these need to be coordinated with rules
+
+             XTOP = $(TOP)
+           BINDIR = /usr/X11R6/bin
+     BUILDINCROOT = $(TOP)/exports
+      BUILDINCDIR = $(BUILDINCROOT)/include
+      BUILDINCTOP = ../..
+      BUILDLIBDIR = $(TOP)/exports/lib
+      BUILDLIBTOP = ../..
+      BUILDBINDIR = $(TOP)/exports/bin
+      BUILDBINTOP = ../..
+   BUILDMODULEDIR = $(BUILDLIBDIR)/modules
+   BUILDMODULETOP = $(BUILDLIBTOP)/..
+    XBUILDINCROOT = $(XTOP)/exports
+     XBUILDINCDIR = $(XBUILDINCROOT)/include/X11
+     XBUILDINCTOP = ../../..
+     XBUILDBINDIR = $(XBUILDINCROOT)/bin
+           INCDIR = $(INCROOT)
+           ADMDIR = /usr/adm
+           LIBDIR = $(USRLIBDIR)/X11
+       LIBEXECDIR = /usr/X11R6/libexec
+        MODULEDIR = $(USRLIBDIR)/modules
+   TOP_X_INCLUDES = -I$(TOP)/exports/include
+
+           VARDIR = /var/X11
+
+        ETCX11DIR = /etc/X11
+
+          CONFDIR = $(ETCX11DIR)
+
+           DOCDIR = $(LIBDIR)/doc
+       DOCHTMLDIR = $(DOCDIR)/html
+         DOCPSDIR = $(DOCDIR)/PostScript
+          FONTDIR = $(LIBDIR)/fonts
+     ENCODINGSDIR = $(FONTDIR)/encodings
+         XINITDIR = $(LIBDIR)/xinit
+           XDMDIR = $(LIBDIR)/xdm
+        XDMVARDIR = $(VARLIBDIR)/xdm
+           TWMDIR = $(LIBDIR)/twm
+           XSMDIR = $(LIBDIR)/xsm
+           NLSDIR = $(LIBDIR)/nls
+       XLOCALEDIR = $(LIBDIR)/locale
+        PEXAPIDIR = $(LIBDIR)/PEX
+      LBXPROXYDIR = $(LIBDIR)/lbxproxy
+  PROXYMANAGERDIR = $(LIBDIR)/proxymngr
+        XPRINTDIR = $(LIBDIR)/xserver
+      XAPPLOADDIR = $(LIBDIR)/app-defaults
+       FONTCFLAGS = -t
+
+     INSTAPPFLAGS = $(INSTDATFLAGS)
+
+              RGB = $(CLIENTENVSETUP) $(XBUILDBINDIR)/rgb
+            FONTC = $(CLIENTENVSETUP) $(XBUILDBINDIR)/bdftopcf
+        MKFONTDIR = $(CLIENTENVSETUP) $(XBUILDBINDIR)/mkfontdir
+
+       DOCUTILSRC = $(XTOP)/doc/util
+        CLIENTSRC = $(TOP)/clients
+          DEMOSRC = $(TOP)/demos
+       XDOCMACROS = $(DOCUTILSRC)/macros.t
+       XIDXMACROS = $(DOCUTILSRC)/indexmacros.t
+       PROGRAMSRC = $(TOP)/programs
+           LIBSRC = $(XTOP)/lib
+          FONTSRC = $(XTOP)/fonts
+     ENCODINGSSRC = $(FONTSRC)/encodings
+       INCLUDESRC = $(BUILDINCROOT)/include
+      XINCLUDESRC = $(INCLUDESRC)/X11
+        SERVERSRC = $(XTOP)/programs/Xserver
+       CONTRIBSRC = $(XTOP)/../contrib
+   UNSUPPORTEDSRC = $(XTOP)/unsupported
+           DOCSRC = $(XTOP)/doc
+           RGBSRC = $(XTOP)/programs/rgb
+      BDFTOPCFSRC = $(PROGRAMSRC)/bdftopcf
+     MKFONTDIRSRC = $(PROGRAMSRC)/mkfontdir
+    FONTSERVERSRC = $(PROGRAMSRC)/xfs
+       FONTINCSRC = $(XTOP)/include/fonts
+        EXTINCSRC = $(XTOP)/include/extensions
+      FTSOURCEDIR = $(TOP)/extras/FreeType
+     XTTSOURCEDIR = $(TOP)/extras/X-TrueType
+     TRANSCOMMSRC = $(LIBSRC)/xtrans
+   TRANS_INCLUDES = -I$(TRANSCOMMSRC)
+ CONNECTION_FLAGS = -DUNIXCONN -DTCPCONN $(STICKY_DEFINES) $(FCHOWN_DEFINES)
+
+       XENVLIBDIR = $(TOP)/exports/lib
+   CLIENTENVSETUP = LD_LIBRARY_PATH=$(XENVLIBDIR)
+
+# $TOG: lnxLib.tmpl /main/7 1998/03/20 14:26:41 mgreess $
+# $XFree86: xc/config/cf/lnxLib.tmpl,v 3.12 1999/06/27 14:07:17 dawes Exp $
+
+          XLIBSRC = $(LIBSRC)/X11
+
+SOXLIBREV = 6.1
+DEPXONLYLIB =
+XONLYLIB =  -lX11
+
+LINTXONLY = $(XLIBSRC)/llib-X11.ln
+
+         XLIBONLY = $(XONLYLIB)
+
+      XEXTLIBSRC = $(LIBSRC)/Xext
+
+SOXEXTREV = 6.4
+DEPEXTENSIONLIB =
+EXTENSIONLIB =  -lXext
+
+LINTEXTENSION = $(XEXTLIBSRC)/llib-Xext.ln
+
+LINTEXTENSIONLIB = $(LINTEXTENSION)
+          DEPXLIB = $(DEPEXTENSIONLIB) $(DEPXONLYLIB)
+             XLIB = $(EXTENSIONLIB) $(XONLYLIB)
+         LINTXLIB = $(LINTXONLYLIB)
+
+    XSSLIBSRC = $(LIBSRC)/Xss
+
+DEPXSSLIB = $(TOP)/exports/lib/libXss.a
+XSSLIB =  -lXss
+
+LINTXSS = $(XSSLIBSRC)/llib-Xss.ln
+
+    XXF86MISCLIBSRC = $(LIBSRC)/Xxf86misc
+
+DEPXXF86MISCLIB = $(TOP)/exports/lib/libXxf86misc.a
+XXF86MISCLIB =  -lXxf86misc
+
+LINTXXF86MISC = $(XXF86MISCLIBSRC)/llib-Xxf86misc.ln
+
+    XXF86VMLIBSRC = $(LIBSRC)/Xxf86vm
+
+DEPXXF86VMLIB = $(TOP)/exports/lib/libXxf86vm.a
+XXF86VMLIB =  -lXxf86vm
+
+LINTXXF86VM = $(XXF86VMLIBSRC)/llib-Xxf86vm.ln
+
+    XXF86DGALIBSRC = $(LIBSRC)/Xxf86dga
+
+DEPXXF86DGALIB = $(TOP)/exports/lib/libXxf86dga.a
+XXF86DGALIB =  -lXxf86dga
+
+LINTXXF86DGA = $(XXF86DGALIBSRC)/llib-Xxf86dga.ln
+
+    XXF86RUSHLIBSRC = $(LIBSRC)/Xxf86rush
+
+DEPXXF86RUSHLIB = $(TOP)/exports/lib/libXxf86rush.a
+XXF86RUSHLIB =  -lXxf86rush
+
+LINTXXF86RUSH = $(XXF86RUSHLIBSRC)/llib-Xxf86rush.ln
+
+    XVLIBSRC = $(LIBSRC)/Xv
+
+DEPXVLIB = $(TOP)/exports/lib/libXv.a
+XVLIB =  -lXv
+
+LINTXV = $(XVLIBSRC)/llib-Xv.ln
+
+    DPSLIBSRC = $(LIBSRC)/dps
+
+SODPSREV = 1.0
+DEPDPSLIB =
+DPSLIB =  -ldps
+
+LINTDPS = $(DPSLIBSRC)/llib-dps.ln
+
+    DPSTKLIBSRC = $(LIBSRC)/dpstk
+
+SODPSTKREV = 1.0
+DEPDPSTKLIB =
+DPSTKLIB =  -ldpstk
+
+LINTDPSTK = $(DPSTKLIBSRC)/llib-dpstk.ln
+
+    PSRESLIBSRC = $(LIBSRC)/psres
+
+SOPSRESREV = 1.0
+DEPPSRESLIB =
+PSRESLIB =  -lpsres
+
+LINTPSRES = $(PSRESLIBSRC)/llib-psres.ln
+
+    GLXLIBSRC = $(LIBSRC)/GL
+
+SOGLREV = 1.2
+DEPGLXLIB =
+GLXLIB =  -lGL
+
+LINTGLX = $(GLXLIBSRC)/llib-GL.ln
+
+    XFONTCACHELIBSRC = $(LIBSRC)/Xfontcache
+
+DEPXFONTCACHELIB = $(TOP)/exports/lib/libXfontcache.a
+XFONTCACHELIB =  -lXfontcache
+
+LINTXFONTCACHE = $(XFONTCACHELIBSRC)/llib-Xfontcache.ln
+
+         XAUTHSRC = $(LIBSRC)/Xau
+
+DEPXAUTHLIB = $(TOP)/exports/lib/libXau.a
+XAUTHLIB =  -lXau
+
+LINTXAUTH = $(XAUTHSRC)/llib-Xau.ln
+
+      XDMCPLIBSRC = $(LIBSRC)/Xdmcp
+
+DEPXDMCPLIB = $(TOP)/exports/lib/libXdmcp.a
+XDMCPLIB =  -lXdmcp
+
+LINTXDMCP = $(XDMCPLIBSRC)/llib-Xdmcp.ln
+
+           XMUSRC = $(LIBSRC)/Xmu
+
+SOXMUREV = 6.1
+DEPXMULIB =
+XMULIB =  -lXmu
+
+LINTXMU = $(XMUSRC)/llib-Xmu.ln
+
+       OLDXLIBSRC = $(LIBSRC)/oldX
+
+DEPOLDXLIB = $(TOP)/exports/lib/liboldX.a
+OLDXLIB =  -loldX
+
+LINTOLDX = $(OLDXLIBSRC)/llib-oldX.ln
+
+         XPLIBSRC = $(LIBSRC)/Xp
+
+SOXPREV = 6.2
+DEPXPLIB =
+XPLIB =  -lXp
+
+LINTXP = $(XPLIBSRC)/llib-Xp.ln
+
+       TOOLKITSRC = $(LIBSRC)/Xt
+
+SOXTREV = 6.0
+DEPXTOOLONLYLIB =
+XTOOLONLYLIB =  -lXt
+
+LINTXTOOLONLY = $(TOOLKITSRC)/llib-Xt.ln
+
+      DEPXTOOLLIB = $(DEPXTOOLONLYLIB) $(DEPSMLIB) $(DEPICELIB)
+         XTOOLLIB = $(XTOOLONLYLIB) $(SMLIB) $(ICELIB)
+     LINTXTOOLLIB = $(LINTXTOOLONLYLIB)
+
+       XALIBSRC = $(LIBSRC)/Xa
+
+SOXAREV = 1.0
+DEPXALIB =
+XALIB =  -lXa
+
+LINTXA = $(XALIBSRC)/llib-Xa.ln
+
+       AWIDGETSRC = $(LIBSRC)/Xaw
+
+SOXAWREV = 7.0
+DEPXAWLIB =
+XAWLIB =  -lXaw
+
+LINTXAW = $(AWIDGETSRC)/llib-Xaw.ln
+
+       AWIDGET6SRC = $(LIBSRC)/Xaw6
+
+SOXAW6REV = 6.1
+DEPXAW6LIB =
+XAW6LIB =  -lXaw
+
+LINTXAW6 = $(AWIDGET6SRC)/llib-Xaw.ln
+
+         XILIBSRC = $(LIBSRC)/Xi
+
+SOXINPUTREV = 6.0
+DEPXILIB =
+XILIB =  -lXi
+
+LINTXI = $(XILIBSRC)/llib-Xi.ln
+
+      XTESTLIBSRC = $(LIBSRC)/Xtst
+
+SOXTESTREV = 6.1
+DEPXTESTLIB =
+XTESTLIB =  -lXtst
+
+LINTXTEST = $(XTESTLIBSRC)/llib-Xtst.ln
+
+        PEXLIBSRC = $(LIBSRC)/PEX5
+
+SOPEXREV = 6.0
+DEPPEXLIB =
+PEXLIB =  -lPEX5
+
+LINTPEX = $(PEXLIBSRC)/llib-PEX5.ln
+
+        XIELIBSRC = $(LIBSRC)/XIE
+
+SOXIEREV = 6.0
+DEPXIELIB =
+XIELIB =  -lXIE
+
+LINTXIE = $(XIELIBSRC)/llib-XIE.ln
+
+      PHIGSLIBSRC = $(LIBSRC)/PHIGS
+
+DEPPHIGSLIB = $(TOP)/exports/lib/libphigs.a
+PHIGSLIB =  -lphigs
+
+LINTPHIGS = $(PHIGSLIBSRC)/llib-phigs.ln
+
+DEPXBSDLIB = $(TOP)/exports/lib/libXbsd.a
+XBSDLIB =  -lXbsd
+
+LINTXBSD = $(LIBSRC)/Xbsd/llib-Xbsd.ln
+
+           ICESRC = $(LIBSRC)/ICE
+
+SOICEREV = 6.3
+DEPICELIB =
+ICELIB =  -lICE
+
+LINTICE = $(ICESRC)/llib-ICE.ln
+
+            SMSRC = $(LIBSRC)/SM
+
+SOSMREV = 6.0
+DEPSMLIB =
+SMLIB =  -lSM
+
+LINTSM = $(SMSRC)/llib-SM.ln
+
+           XKEYSRC = $(LIBSRC)/Xkey
+
+SOXKEYREV = 6.0
+DEPXKEYLIB =
+XKEYLIB =  -lXkey
+
+LINTXKEY = $(XKEYSRC)/llib-Xkey.ln
+
+         FSLIBSRC = $(LIBSRC)/FS
+
+DEPFSLIB = $(TOP)/exports/lib/libFS.a
+FSLIB =  -lFS
+
+LINTFS = $(FSLIBSRC)/llib-FS.ln
+
+         FONTLIBSRC = $(LIBSRC)/font
+
+SOFONTREV = 1.3
+DEPFONTLIB =
+FONTLIB = -L$(FREETYPELIBDIR) -L$(FONTLIBSRC)  -lXfont
+
+LINTXFONT = $(FONTLIBSRC)/llib-Xfont.ln
+#
+SOFONTREV = 1.3
+DEPXFONTLIB =
+XFONTLIB =  -lXfont
+
+LINTXFONT = $(FONTLIBSRC)/llib-Xfont.ln
+
+     FONTSTUBLIBSRC = $(FONTLIBSRC)/stubs
+
+DEPFONTSTUBLIB = $(TOP)/exports/lib/libfntstubs.a
+FONTSTUBLIB =  -lfntstubs
+
+LINTFONTSTUB = $(FONTSUBLIBSRC)/llib-fntstubs.ln
+         DEPFONTLIB = $(DEPXFONTLIB) $(DEPFONTSTUBLIB)
+            FONTLIB = $(XFONTLIB) $(FONTSTUBLIB)
+
+          XPMLIBSRC = $(LIBSRC)/Xpm
+
+SOXPMREV = 4.11
+DEPXPMLIB =
+XPMLIB =  -lXpm
+
+LINTXPM = $(XPMLIBSRC)/llib-Xpm.ln
+
+    XKBFILELIBSRC = $(LIBSRC)/xkbfile
+
+DEPXKBFILELIB = $(TOP)/exports/lib/libxkbfile.a
+XKBFILELIB =  -lxkbfile
+
+LINTXKBFILE = $(XKBFILELIBSRC)/llib-xkbfile.ln
+
+     XKBCOMPSRC = $(PROGRAMSRC)/xkbcomp
+     XKBCOMPCMD = $(CLIENTENVSETUP) $(XBUILDBINDIR)/xkbcomp
+
+    XKBUILIBSRC = $(LIBSRC)/xkbui
+
+DEPXKBUILIB = $(TOP)/exports/lib/libxkbui.a
+XKBUILIB =  -lxkbui
+
+LINTXKBUI = $(XKBUILIBSRC)/llib-xkbui.ln
+
+          DEPLIBS = $(DEPXAWLIB) $(DEPXMULIB) $(DEPXTOOLLIB) $(DEPXLIB)
+
+         DEPLIBS1 = $(DEPLIBS)
+         DEPLIBS2 = $(DEPLIBS)
+         DEPLIBS3 = $(DEPLIBS)
+         DEPLIBS4 = $(DEPLIBS)
+         DEPLIBS5 = $(DEPLIBS)
+         DEPLIBS6 = $(DEPLIBS)
+         DEPLIBS7 = $(DEPLIBS)
+         DEPLIBS8 = $(DEPLIBS)
+         DEPLIBS9 = $(DEPLIBS)
+         DEPLIBS10 = $(DEPLIBS)
+
+XMULIBONLY = -lXmu
+XMULIB = $(XMULIBONLY) $(XTOOLLIB) $(XLIB)
+
+        CONFIGDIR = $(LIBDIR)/config
+
+    USRLIBDIRPATH = $(USRLIBDIR)
+        LDPRELIBS = -L$(BUILDLIBDIR)
+       LDPOSTLIBS =
+     TOP_INCLUDES = -I$(TOP) $(TOP_X_INCLUDES)
+  PROJECT_DEFINES =
+
+CXXPROJECT_DEFINES =
+
+# ----------------------------------------------------------------------
+# start of Imakefile
+
+# $TOG: Imakefile /main/27 1997/04/30 15:38:09 kaleb $
+#
+#
+#
+#
+# $XFree86: xc/config/makedepend/Imakefile,v 3.7 1999/08/14 10:49:02 dawes Exp $
+
+    SRCS = include.c main.c parse.c pr.c cppsetup.c ifparser.c
+    OBJS = include.o main.o parse.o pr.o cppsetup.o ifparser.o
+
+# Some compilers generate fatal errors if an -L directory does
+# not exist.  Since BUILDLIBDIR may not exist yet suppress its use.
+         LDPRELIB =
+        LDPRELIBS =
+
+PREINCDIR = /usr/local/lib/gcc-include
+PREINC_DEFINES = -DPREINCDIR=\"$(PREINCDIR)\"
+
+STDINCDIR = /usr/include
+STDINC_DEFINES = -DINCLUDEDIR=\"$(STDINCDIR)\"
+
+POSTINCDIR = /usr/lib/gcc-lib/i586-pld-linux/2.95.2/include
+POSTINC_DEFINES = -DPOSTINCDIR=\"$(POSTINCDIR)\"
+
+INC_DEFINES = $(PREINC_DEFINES) $(STDINC_DEFINES) $(POSTINC_DEFINES)
+OSUF = .o
+OSUF_DEFINES = -DOBJSUFFIX=\"$(OSUF)\"
+MAIN_DEFINES = $(INC_DEFINES) $(SIGNAL_DEFINES) $(OSUF_DEFINES)
+INCLUDES = -I$(IMAKESRC) -I$(TOP)/include $(TOP_X_INCLUDES)/X11
+
+ DEPLIBS =
+
+all:: makedepend
+
+bootstrap:
+
+clean::
+	$(RM) Makefile.proto
+
+ PROGRAM = makedepend
+
+all:: makedepend
+
+makedepend: $(OBJS) $(DEPLIBS)
+	$(RM) $@
+	$(CCLINK) -o $@ $(LDOPTIONS) $(OBJS) $(LOCAL_LIBRARIES) $(LDLIBS)  $(EXTRA_LOAD_FLAGS)
+
+install:: makedepend
+	@if [ -d $(DESTDIR)$(BINDIR) ]; then set +x; \
+	else (set -x; $(MKDIRHIER) $(DESTDIR)$(BINDIR)); fi
+	$(INSTALL) $(INSTALLFLAGS) $(INSTPGMFLAGS)  makedepend $(DESTDIR)$(BINDIR)/makedepend
+
+depend:: $(DEPEND)
+
+$(DEPEND):
+	@echo "checking $@ over in $(DEPENDSRC) first..."; \
+	cd $(DEPENDSRC) && $(MAKE); \
+	echo "okay, continuing in $(CURRENT_DIR)"
+
+depend::
+	$(DEPEND) $(DEPENDFLAGS) -- $(ALLDEFINES) $(DEPEND_DEFINES) -- $(SRCS)
+
+lint:
+	$(LINT) $(LINTFLAGS) $(SRCS) $(LINTLIBS)
+lint1:
+	$(LINT) $(LINTFLAGS) $(FILE) $(LINTLIBS)
+
+clean::
+	$(RM) makedepend
+
+main.o:	main.c $(ICONFIGFILES)
+	$(RM) $@
+	 $(CC) -c $(CFLAGS) $(MAIN_DEFINES) $*.c
+
+main.i: main.c $(ICONFIGFILES)
+	$(RM) $@
+	 $(CC) -E $(CFLAGS) $(MAIN_DEFINES) main.c > $@
+
+main.s: main.c $(ICONFIGFILES)
+	$(RM) $@
+	 $(CC) -S $(CFLAGS) $(MAIN_DEFINES) $*.c
+
+install.man:: mkdepend.man
+	@if [ -d $(DESTDIR)$(MANDIR) ]; then set +x; \
+	else (set -x; $(MKDIRHIER) $(DESTDIR)$(MANDIR)); fi
+	$(INSTALL) $(INSTALLFLAGS) $(INSTMANFLAGS) mkdepend.man $(DESTDIR)$(MANDIR)/makedepend.$(MANSUFFIX)
+
+install.sdk:: makedepend
+	@if [ -d $(DESTDIR)$(DRIVERSDKDIR)/config/makedepend ]; then set +x; \
+	else (set -x; $(MKDIRHIER) $(DESTDIR)$(DRIVERSDKDIR)/config/makedepend); fi
+	$(INSTALL) -c $(INSTBINFLAGS) makedepend $(DESTDIR)$(DRIVERSDKDIR)/config/makedepend/makedepend
+
+# ----------------------------------------------------------------------
+# common rules for all Makefiles - do not edit
+
+.c.i:
+	$(RM) $@
+	 $(CC) -E $(CFLAGS) $(_NOOP_) $*.c > $@
+
+.SUFFIXES: .s
+
+.c.s:
+	$(RM) $@
+	 $(CC) -S $(CFLAGS) $(_NOOP_) $*.c
+
+emptyrule::
+
+clean::
+	$(RM) *.CKP *.ln *.BAK *.bak *.o core errs ,* *~ *.a .emacs_* tags TAGS make.log MakeOut  "#"*
+
+Makefile:: $(IMAKE)
+
+$(IMAKE) $(IMAKE).o:
+	-@(cd $(IMAKESRC) && if [ -f Makefile ]; then \
+	echo "checking $@ in $(IMAKESRC) first..."; $(MAKE) all; else \
+	echo "bootstrapping $@ from Makefile.ini in $(IMAKESRC) first..."; \
+	$(MAKE) -f Makefile.ini BOOTSTRAPCFLAGS="$(BOOTSTRAPCFLAGS)"; fi; \
+	echo "okay, continuing in $(CURRENT_DIR)")
+
+Makefile::
+	-@if [ -f Makefile ]; then set -x; \
+	$(RM) Makefile.bak; $(MV) Makefile Makefile.bak; \
+	else exit 0; fi
+	$(IMAKE_CMD) -DTOPDIR=$(TOP) -DCURDIR=$(CURRENT_DIR)
+
+tags::
+	$(TAGS) -w *.[ch]
+	$(TAGS) -xw *.[ch] > TAGS
+
+man_keywords::
+
+# ----------------------------------------------------------------------
+# empty rules for directories that do not have SUBDIRS - do not edit
+
+install::
+	@echo "install in $(CURRENT_DIR) done"
+
+install.man::
+	@echo "install.man in $(CURRENT_DIR) done"
+
+install.sdk::
+	@echo "install.sdk in $(CURRENT_DIR) done"
+
+Makefiles::
+
+includes::
+
+depend::
+
+# ----------------------------------------------------------------------
+# dependencies generated by makedepend
+
diff -urN XFree86-4.0.orig/xc/programs/Xserver/hw/xfree86/XF86Setup/Imakefile XFree86-4.0/xc/programs/Xserver/hw/xfree86/XF86Setup/Imakefile
--- XFree86-4.0.orig/xc/programs/Xserver/hw/xfree86/XF86Setup/Imakefile	Mon Sep  6 15:23:12 1999
+++ XFree86-4.0/xc/programs/Xserver/hw/xfree86/XF86Setup/Imakefile	Fri Mar 17 02:19:49 2000
@@ -118,7 +118,7 @@
 
  XCONFIG_DEFINES = $(EXT_DEFINES) -DXF86SETUP \
 		   '-DPROJECTROOT="$(PROJECTROOT)"' \
-		   '-DSERVER_CONFIG_FILE="$(LIBDIR)/$(XCONFIGFILE)"' \
+		   '-DSERVER_CONFIG_FILE="/etc/X11/$(XCONFIGFILE)"' \
 		   $(SITE_FONT_PATH) $(SITE_RGB_DB) \
 		   -DNO_WRAPPERS
 
diff -urN XFree86-4.0.orig/xc/programs/Xserver/hw/xfree86/XF86Setup/phase5.tcl XFree86-4.0/xc/programs/Xserver/hw/xfree86/XF86Setup/phase5.tcl
--- XFree86-4.0.orig/xc/programs/Xserver/hw/xfree86/XF86Setup/phase5.tcl	Mon Apr  6 00:30:46 1998
+++ XFree86-4.0/xc/programs/Xserver/hw/xfree86/XF86Setup/phase5.tcl	Fri Mar 17 02:19:49 2000
@@ -38,7 +38,7 @@
 	set devid [lindex $DeviceIDs $idx]
         global Device_$devid
         set server [set Device_${devid}(Server)]
-	set linkname $Xwinhome/bin/X
+	set linkname /etc/X11/X
 	set lastlink $linkname
 	for {set nlinks 0} \
 		{[file exists $linkname] && [file type $linkname]=="link" \
@@ -74,7 +74,7 @@
 	    set CWD [pwd]
 	    cd $linkdir
 	    catch "unlink X" ret
-	    if [catch "link $Xwinhome/bin/XF86_$server X" ret] {
+	    if [catch "link ../../usr/X11R6/bin/XF86_$server X" ret] {
 		mesg $messages(phase5.4) okay
 	    } else {
 		mesg $messages(phase5.5) okay
diff -urN XFree86-4.0.orig/xc/programs/xinit/Imakefile XFree86-4.0/xc/programs/xinit/Imakefile
--- XFree86-4.0.orig/xc/programs/xinit/Imakefile	Wed Feb  9 19:02:36 2000
+++ XFree86-4.0/xc/programs/xinit/Imakefile	Fri Mar 17 02:19:49 2000
@@ -39,11 +39,13 @@
 InstallNamedProg(startx,startx,$(BINDIR))
 InstallManPage(startx,$(MANDIR))
 
-#if InstallXinitConfig
-InstallNonExecFile($(SAMPLECONFIG),$(XINITDIR))
-#else
-InstallNonExecFileNoClobber($(SAMPLECONFIG),$(XINITDIR))
-#endif
+XCOMM Red Hat gets this from a different package
+XCOMM #if InstallXinitConfig
+XCOMM InstallNonExecFile($(SAMPLECONFIG),$(XINITDIR))
+XCOMM #else
+XCOMM InstallNonExecFileNoClobber($(SAMPLECONFIG),$(XINITDIR))
+XCOMM #endif
+
 #endif
 
 LinkConfDirectory(xinit,.,xinit,.)
diff -urN XFree86-4.0.orig/xc/programs/xterm/Imakefile XFree86-4.0/xc/programs/xterm/Imakefile
--- XFree86-4.0.orig/xc/programs/xterm/Imakefile	Sat Mar  4 05:42:24 2000
+++ XFree86-4.0/xc/programs/xterm/Imakefile	Fri Mar 17 02:19:49 2000
@@ -36,6 +36,10 @@
 #define XkbClientLibs /**/
 #endif
 
+#if InstallXtermSetUID
+#undef InstallXtermSetUID
+#endif
+
 #ifndef SpecialCObjectRule
 #define SpecialCObjectRule(module,ignore,defines) \
 module.o: ; $(CC) -c defines $(CFLAGS) module.c
@@ -57,6 +61,8 @@
          PTYLIB = -lpucc
 #endif
 
+         PTYLIB = -lutil
+
 #if ((LinuxCLibMajorVersion == 6) && (LinuxCLibMinorVersion < 1))
          PTYLIB = -lutil
 #endif
@@ -101,7 +107,7 @@
 		  -DOSMINORVERSION=$(OSMINORVERSION)
    MISC_DEFINES = /* -DALLOWLOGGING -DALLOWLOGFILEEXEC */
     XKB_DEFINES = XkbClientDefines
-        DEFINES = -I. $(XKB_DEFINES) $(TERMCAPDEFINES) $(FEATURE_DEFINES) $(SCROLLBAR_RIGHT) $(UTF8_OPTION)
+        DEFINES = -I. -I/usr/include/ncurses $(XKB_DEFINES) $(TERMCAPDEFINES) $(FEATURE_DEFINES) $(SCROLLBAR_RIGHT) $(UTF8_OPTION)
 
 #ifdef OS2Architecture
           MAINSRC = os2main.c
diff -urN XFree86-4.0.orig/xc/programs/xterm/xterm.h XFree86-4.0/xc/programs/xterm/xterm.h
--- XFree86-4.0.orig/xc/programs/xterm/xterm.h	Wed Mar  1 02:21:05 2000
+++ XFree86-4.0/xc/programs/xterm/xterm.h	Fri Mar 17 02:19:49 2000
@@ -58,7 +58,7 @@
 #endif
 
 #ifndef DFT_TERMTYPE
-#define DFT_TERMTYPE "xterm"
+#define DFT_TERMTYPE "xterm-color"
 #endif
 
 #ifndef X_NOT_POSIX
