--- xc/programs/xterm/Imakefile.orig	2007-12-02 16:41:43.351522957 +0100
+++ xc/programs/xterm/Imakefile	2007-12-02 16:43:16.500831234 +0100
@@ -50,6 +50,8 @@
         UTMPLIB = -lutempter
 #endif
 
+COLORS_256 = -DOPT_256_COLORS=1
+
 /*
  * Fixes to allow compile with X11R5, etc.
  */
@@ -209,7 +211,7 @@
    PATH_DEFINES = -DPROJECTROOT=$(PROJECTROOT)
         DEFINES = $(XKB_DEFINES) $(TERMCAPDEFINES) $(FEATURE_DEFINES) \
                   $(SCROLLBAR_RIGHT) $(UTF8_OPTION) $(XRFDEF) $(PATH_DEFINES) \
-                  $(PUTENVDEF) $(IMAKEDEFINES) $(TRACEDEF)
+                  $(PUTENVDEF) $(IMAKEDEFINES) $(TRACEDEF) $(COLORS_256)
        INCLUDES = $(OBSOLETE_INCLUDES) $(XRFINCLUDES) -I/usr/include/ncurses
 
 #ifdef OS2Architecture
