--- XFree86-4.0/xc/config/cf/Imake.rules.shared	Fri May 12 09:52:32 2000
+++ XFree86-4.0/xc/config/cf/Imake.rules	Fri May 12 09:57:14 2000
@@ -3082,6 +3082,7 @@
 #ifndef LibraryModuleTarget
 #if MakeDllModules
 #define LibraryModuleTarget(module,objects)				@@\
+NormalLibraryTarget(module, objects)					@@\
 DynamicModuleTarget(Concat3(lib,module,.so), objects)
 #else
 #define LibraryModuleTarget(module, objects)				@@\
@@ -3095,6 +3096,7 @@
 #ifndef DepLibraryModuleTarget
 #if MakeDllModules
 #define DepLibraryModuleTarget(module,deplist,objects)			@@\
+NormalDepLibraryTarget(module, deplist, objects)			@@\
 DepDynamicModuleTarget(Concat3(lib,module,.so), deplist, objects)
 #else
 #define DepLibraryModuleTarget(module,deplist,objects)			@@\
--- XFree86-4.0/xc/programs/Xserver/GL/Imakefile.shared	Sat Feb 26 00:37:09 2000
+++ XFree86-4.0/xc/programs/Xserver/GL/Imakefile	Fri May 12 01:08:54 2000
@@ -42,7 +42,7 @@
 
 ModuleObjectRule()
 
-NormalDepLibraryTarget(glx,$(SUBDIRS) $(DONES) $(MOBJS),$(OBJS))
+DepLibraryModuleTarget(glx,$(SUBDIRS) $(DONES) $(MOBJS),$(OBJS))
 
 InstallLibraryModule(glx,$(MODULEDIR),extensions)
 ForceSubdirs($(SUBDIRS))
--- XFree86-4.0/xc/programs/Xserver/fb/Imakefile.shared	Sat Feb 26 00:37:40 2000
+++ XFree86-4.0/xc/programs/Xserver/fb/Imakefile	Fri May 12 01:08:54 2000
@@ -79,7 +79,7 @@
 	     ../mi/llib-lmi.ln
 
 NormalLibraryObjectRule()
-NormalLibraryTarget(fb,$(OBJS))
+LibraryModuleTarget(fb,$(OBJS))
 LintLibraryTarget(fb,$(SRCS))
 
 NormalLintTarget($(SRCS))
--- XFree86-4.0/xc/programs/Xserver/Imakefile.shared	Fri May 12 09:27:35 2000
+++ XFree86-4.0/xc/programs/Xserver/Imakefile	Fri May 12 09:30:32 2000
@@ -527,13 +527,13 @@
 #if XF86Server
 XF86SERVER = XFree86
 #endif
-#if XnestServer && !MakeDllModules
+#if XnestServer 
 XNEST = Xnest
 #endif
-#if XVirtualFramebufferServer && !MakeDllModules
+#if XVirtualFramebufferServer
 XVFB = Xvfb
 #endif
-#if defined(XF86Server) || (!MakeDllModules && (defined(XnestServer) || defined(XVirtualFramebufferServer)))
+#if defined(XF86Server) || ((defined(XnestServer) || defined(XVirtualFramebufferServer)))
 MakeMutex($(XF86SERVER) $(XNEST) $(XVFB))
 #endif
 MakeMutex($(STDDIRS) mfb fb cfb cfb16 cfb24 cfb32)
@@ -863,7 +863,7 @@
 KDRIVEDIRS=$(KDDIRS) $(FBDEVDIR) $(SAVAGEDIR) $(TRIDENTDIR) $(SIS530DIR) \
 	$(TRIODIR) $(TS300DIR) $(ITSYDIR)
 
-#if XprtServer && !MakeDllModules
+#if XprtServer
 XCOMM
 XCOMM Print Server
 XCOMM
@@ -896,7 +896,7 @@
 ServerTarget(Xprt,$(XPSUBDIRS),$(XPOBJS),$(XPLIBS) $(LOADABLEEXTS) $(LIBCWRAPPER),$(XPSYSLIBS))
 #endif	/* XprtServer */
 
-#if XnestServer && !MakeDllModules
+#if XnestServer
 XCOMM
 XCOMM Server with Xlib-based ddx
 XCOMM
@@ -945,7 +945,7 @@
 #endif /* XnonServer */
 
 
-#if XVirtualFramebufferServer && !MakeDllModules
+#if XVirtualFramebufferServer
 XCOMM
 XCOMM server with Virtual (malloced) framebuffer
 XCOMM
