--- XFree86-4.0.2/xc/programs/Xserver/hw/xfree86/os-support/linux/lnx_pci.c.orig	Tue Oct 17 20:53:20 2000
+++ XFree86-4.0.2/xc/programs/Xserver/hw/xfree86/os-support/linux/lnx_pci.c	Fri Mar  2 13:23:04 2001
@@ -13,7 +13,7 @@ Bool
 xf86GetPciSizeFromOS(PCITAG tag, int index, int* bits)
 {
     FILE *file;
-    char c[0x100];
+    char c[0x200];
     char *res;
     int bus, devfn, dev, fn;
     unsigned int size[7];
@@ -26,7 +26,7 @@ xf86GetPciSizeFromOS(PCITAG tag, int ind
     if (!(file = fopen("/proc/bus/pci/devices","r")))
 	return FALSE;
     do {
-	res = fgets(c,0xff,file);
+	res = fgets(c,0x1ff,file);
 	if (res) {
 	    num = sscanf(res,"%02x%02x\t%*04x%*04x\t%*x"
 			     "\t%*x\t%*x\t%*x\t%*x\t%*x\t%*x\t%*x"
