diff -urN xc.org/include/Xos.h xc/include/Xos.h
--- xc.org/include/Xos.h	Fri Mar  9 09:04:10 2001
+++ xc/include/Xos.h	Fri Mar  9 10:22:56 2001
@@ -155,6 +155,7 @@
 #ifdef SYSV
 
 #ifndef USL
+#include <time.h>
 #include <sys/time.h>
 #endif
 #include <time.h>
@@ -186,6 +187,7 @@
 #if defined(_POSIX_SOURCE) && defined(SVR4)
 /* need to omit _POSIX_SOURCE in order to get what we want in SVR4 */
 #undef _POSIX_SOURCE
+#include <time.h>
 #include <sys/time.h>
 #define _POSIX_SOURCE
 #elif defined(WIN32)
@@ -218,6 +220,7 @@
 #include <sys/time.h>
 #include <time.h>
 #else
+#include <time.h>
 #include <sys/time.h>
 #endif /* QNX */
 #endif /* defined(_POSIX_SOURCE) && defined(SVR4) */

