diff -urN xc.orig/config/cf/xfree86.cf xc/config/cf/xfree86.cf
--- xc.orig/config/cf/xfree86.cf	Mon Dec 30 15:52:31 2002
+++ xc/config/cf/xfree86.cf	Mon Dec 30 20:44:34 2002
@@ -1122,7 +1122,7 @@
   && !defined(ia64Architecture) \
   && !defined(s390xArchitecture) \
   && !defined (x86_64Architecture)
-#  define BuildPexExt		NO	/* But we turn off PEX anyway now */
+#  define BuildPexExt		YES	/* But we turn off PEX anyway now */
 # else
 #  define BuildPexExt		NO
 # endif
@@ -1363,7 +1363,7 @@
  */
 #ifndef BuildXIE
 # if XF86Server
-#  define BuildXIE		NO
+#  define BuildXIE		YES
 # else
 #  define BuildXIE		NO
 # endif
diff -urN xc.orig/programs/xdpyinfo/Imakefile xc/programs/xdpyinfo/Imakefile
--- xc.orig/programs/xdpyinfo/Imakefile	Mon Dec 30 15:54:07 2002
+++ xc/programs/xdpyinfo/Imakefile	Mon Dec 30 20:40:08 2002
@@ -10,11 +10,6 @@
         SHMDEFINES = -DMITSHM
 #endif
 
-#if BuildXIElib
-        XIEDEFINES = -DXIE
-        XIEDEPLIBS = $(DEPXIELIB)
-           XIELIBS = $(XIELIB)
-#endif
 
 #if BuildXF86VidModeLibrary
     VIDMODEDEFINES = -DXF86VIDMODE
