--- xc/config/cf/site.def~	Wed Feb 27 01:51:12 2002
+++ xc/config/cf/site.def	Sat Jan 11 12:44:33 2003
@@ -72,9 +72,10 @@
 #ifdef AfterVendorCF
 
 #ifndef ProjectRoot
-#define ProjectRoot /usr/X11R6
+#define ProjectRoot /usr
 #endif
 
+#define ManDirectoryRoot /usr/share/man
 /*
  * On some platforms, some things may be installed outside of ProjectRoot
  * by default.  To disable this, uncomment the following line.
